// @reatom/primitives/build/index.js
"use strict";
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
const __reatom_core = __toESM(require("@reatom/core"));
const __reatom_utils = __toESM(require("@reatom/utils"));

//#region src/withAssign.ts
const withAssign = (getProps) => (target) => (0, __reatom_utils.assign)(target, getProps(target, target.__reatom.name));

//#endregion
//#region src/reatomArray.ts
const reatomArray = (initState = [], name) => (0, __reatom_core.atom)(initState, name).pipe(withAssign((target, name$1) => ({
	__reatomArray: true,
	toReversed: (0, __reatom_core.action)((ctx) => target(ctx, (prev) => prev.slice().reverse()), `${name$1}.toReversed`),
	toSorted: (0, __reatom_core.action)((ctx, compareFn) => target(ctx, (prev) => prev.slice().sort(compareFn)), `${name$1}.toSorted`),
	toSpliced: (0, __reatom_core.action)((ctx, start, deleteCount, ...items) => target(ctx, (state) => {
		state = state.slice();
		state.splice(start, deleteCount, ...items);
		return state;
	}), `${name$1}.toSpliced`),
	with: (0, __reatom_core.action)((ctx, i, value) => target(ctx, (state) => {
		if (Object.is(state.at(i), value)) return state;
		state = state.slice();
		state[i] = value;
		return state;
	}), `${name$1}.with`),
	push: (0, __reatom_core.action)((ctx, ...items) => {
		const arrCopy = ctx.get(target).slice();
		const pushed = arrCopy.push(...items);
		target(ctx, arrCopy);
		return pushed;
	}, `${name$1}.push`),
	pop: (0, __reatom_core.action)((ctx) => {
		const arrCopy = ctx.get(target).slice();
		const popped = arrCopy.pop();
		target(ctx, arrCopy);
		return popped;
	}, `${name$1}.pop`),
	shift: (0, __reatom_core.action)((ctx) => {
		const arrCopy = ctx.get(target).slice();
		const shifted = arrCopy.shift();
		target(ctx, arrCopy);
		return shifted;
	}, `${name$1}.shift`),
	unshift: (0, __reatom_core.action)((ctx, ...items) => {
		const arrCopy = ctx.get(target).slice();
		const unshifted = arrCopy.unshift(...items);
		target(ctx, arrCopy);
		return unshifted;
	}, `${name$1}.unshift`)
})));
const isArrayAtom = (thing) => thing?.__reatomArray === true;

//#endregion
//#region src/reatomBoolean.ts
const reatomBoolean = (init = false, name) => (0, __reatom_core.atom)(init, name).pipe(withAssign((target, name$1) => ({
	toggle: (0, __reatom_core.action)((ctx) => target(ctx, (prev) => !prev), `${name$1}.toggle`),
	setTrue: (0, __reatom_core.action)((ctx) => target(ctx, true), `${name$1}.setTrue`),
	setFalse: (0, __reatom_core.action)((ctx) => target(ctx, false), `${name$1}.setFalse`),
	reset: (0, __reatom_core.action)((ctx) => target(ctx, init), `${name$1}.reset`)
})));

//#endregion
//#region src/reatomEnum.ts
const reatomEnum = (variants, options = {}) => {
	const { name, format = "camelCase", initState = variants[0] } = typeof options === "string" ? { name: options } : options;
	const stateAtom = (0, __reatom_core.atom)(initState, name);
	const enumAtom = Object.assign((ctx, update) => {
		const state = stateAtom(ctx, update);
		(0, __reatom_core.throwReatomError)(!variants.includes(state), `invalid enum value "${state}" for "${name}" enum`);
		return state;
	}, stateAtom);
	const cases = enumAtom.enum = {};
	enumAtom.reset = (0, __reatom_core.action)((ctx) => enumAtom(ctx, initState), `${name}.reset`);
	for (const variant of variants) {
		cases[variant] = variant;
		const setterName = variant.replace(/^./, (firstLetter) => "set" + (format === "camelCase" ? firstLetter.toUpperCase() : `_${firstLetter}`));
		enumAtom[setterName] = (0, __reatom_core.action)((ctx) => enumAtom(ctx, variant), `${name}.${setterName}`);
	}
	return enumAtom;
};

//#endregion
//#region src/reatomMap.ts
const reatomMap = (initState = new Map(), name) => {
	const atomInitState = initState instanceof Map ? initState : new Map(initState);
	return (0, __reatom_core.atom)(atomInitState, name).pipe(withAssign((target, name$1) => {
		const getOrCreate = (0, __reatom_core.action)((ctx, key, value) => {
			actions.set(ctx, key, value);
			return value;
		}, `${name$1}.getOrCreate`);
		const actions = {
			get: (ctx, key) => ctx.get(target).get(key),
			getOrCreate: (ctx, key, creator) => actions.has(ctx, key) ? actions.get(ctx, key) : getOrCreate(ctx, key, creator()),
			has: (ctx, key) => ctx.get(target).has(key),
			set: (0, __reatom_core.action)((ctx, key, value) => target(ctx, (prev) => {
				const valuePrev = prev.get(key);
				return Object.is(valuePrev, value) && (value !== undefined || prev.has(key)) ? prev : new Map(prev).set(key, value);
			}), `${name$1}.set`),
			delete: (0, __reatom_core.action)((ctx, key) => target(ctx, (prev) => {
				if (!prev.has(key)) return prev;
				const next = new Map(prev);
				next.delete(key);
				return next;
			}), `${name$1}.delete`),
			clear: (0, __reatom_core.action)((ctx) => target(ctx, new Map()), `${name$1}.clear`),
			reset: (0, __reatom_core.action)((ctx) => target(ctx, atomInitState), `${name$1}.reset`),
			sizeAtom: (0, __reatom_core.atom)((ctx) => ctx.spy(target).size, `${name$1}.size`)
		};
		return actions;
	}));
};

//#endregion
//#region src/reatomLinkedList.ts
const readonly = (anAtom) => ({ ...anAtom });
const LL_PREV = Symbol("Reatom linked list prev");
const LL_NEXT = Symbol("Reatom linked list next");
const addLL = (state, node, after) => {
	if (node === after) return;
	if (after) {
		node[LL_PREV] = after;
		node[LL_NEXT] = after[LL_NEXT];
		after[LL_NEXT] = node;
		if (state.tail === after) state.tail = node;
	} else {
		node[LL_PREV] = null;
		node[LL_NEXT] = state.head;
		if (state.head) state.head[LL_PREV] = node;
		if (!state.tail) state.tail = node;
		state.head = node;
	}
	state.size++;
};
const removeLL = (state, node) => {
	if (state.head === node) state.head = node[LL_NEXT];
	else if (node[LL_PREV] !== null) node[LL_PREV][LL_NEXT] = node[LL_NEXT];
	if (state.tail === node) state.tail = node[LL_PREV];
	else if (node[LL_NEXT] !== null) node[LL_NEXT][LL_PREV] = node[LL_PREV];
	node[LL_PREV] = null;
	node[LL_NEXT] = null;
	state.size--;
};
const swapLL = (state, a, b) => {
	if (a === b) return;
	if (state.head === b) return swapLL(state, b, a);
	const prevA = a[LL_PREV];
	const nextA = a[LL_NEXT];
	const prevB = b[LL_PREV];
	const nextB = b[LL_NEXT];
	if (nextA === b) {
		a[LL_NEXT] = nextB;
		b[LL_PREV] = prevA;
		b[LL_NEXT] = a;
		a[LL_PREV] = b;
		if (nextB) nextB[LL_PREV] = a;
		if (prevA) prevA[LL_NEXT] = b;
	} else if (nextB === a) {
		b[LL_NEXT] = nextA;
		a[LL_PREV] = prevB;
		a[LL_NEXT] = b;
		b[LL_PREV] = a;
		if (nextA) nextA[LL_PREV] = b;
		if (prevB) prevB[LL_NEXT] = a;
	} else {
		if (prevA) prevA[LL_NEXT] = b;
		if (nextA) nextA[LL_PREV] = b;
		if (prevB) prevB[LL_NEXT] = a;
		if (nextB) nextB[LL_PREV] = a;
		a[LL_PREV] = prevB;
		a[LL_NEXT] = nextB;
		b[LL_PREV] = prevA;
		b[LL_NEXT] = nextA;
	}
	if (state.head === a) state.head = b;
	else if (state.head === b) state.head = a;
	if (state.tail === a) state.tail = b;
	else if (state.tail === b) state.tail = a;
};
const moveLL = (state, node, after) => {
	removeLL(state, node);
	addLL(state, node, after);
};
const clearLL = (state) => {
	while (state.tail) removeLL(state, state.tail);
};
const toArray = (head, prev) => {
	let arr = [];
	let i = 0;
	while (head) {
		if (prev !== undefined && prev[i] !== head) prev = undefined;
		arr.push(head);
		head = head[LL_NEXT];
		i++;
	}
	return arr.length === prev?.length ? prev : arr;
};
function reatomLinkedList(options, name = (0, __reatom_core.__count)("reatomLinkedList")) {
	const { create: userCreate = (ctx, ...params) => params[0], key = undefined,...restOptions } = typeof options === "function" ? { create: options } : Array.isArray(options) ? {
		create: (ctx, ...params) => params[0],
		initState: options
	} : options;
	const _name = name;
	const isLL = (node) => !!node && LL_NEXT in node && LL_PREV in node;
	const throwModel = (node) => (0, __reatom_core.throwReatomError)(isLL(node), "The data is already in a linked list.");
	const throwNotModel = (node) => (0, __reatom_core.throwReatomError)(!isLL(node), "The passed data is not a linked list node.");
	let STATE = null;
	const linkedList = (0, __reatom_core.atom)(STATE, name);
	linkedList.__reatom.initState = (ctx) => {
		try {
			if ("initState" in restOptions) return createLinkedList(restOptions.initState ?? []);
			else if ("initSnapshot" in restOptions) return createLinkedList(ctx, restOptions.initSnapshot ?? []);
			else return createLinkedList([]);
		} finally {
			STATE = null;
		}
	};
	const createLinkedList = (ctxOrInitState, initSnapshot) => {
		const initState = Array.isArray(ctxOrInitState) ? ctxOrInitState : initSnapshot.map((params) => userCreate(ctxOrInitState, ...params));
		const state = {
			size: 0,
			version: 1,
			changes: [],
			head: null,
			tail: null
		};
		for (const node of initState) {
			throwModel(node);
			addLL(state, node, state.tail);
		}
		return state;
	};
	const batchFn = (ctx, cb) => {
		if (STATE) return cb(ctx);
		STATE = linkedList(ctx, ({ head, tail, size, version }) => ({
			size,
			version: version + 1,
			changes: [],
			head,
			tail
		}));
		try {
			return cb(ctx);
		} finally {
			STATE = null;
		}
	};
	const batch = (0, __reatom_core.action)(batchFn, `${name}._batch`);
	const create = (0, __reatom_core.action)((ctx, ...params) => {
		return batchFn(ctx, () => {
			const node = userCreate(ctx, ...params);
			(0, __reatom_core.throwReatomError)(!(0, __reatom_utils.isObject)(node) && typeof node !== "function", `reatomLinkedList can operate only with objects or functions, received "${node}".`);
			throwModel(node);
			addLL(STATE, node, STATE.tail);
			STATE.changes.push({
				kind: "create",
				node
			});
			return node;
		});
	}, `${name}.create`);
	const remove = (0, __reatom_core.action)((ctx, node) => {
		return batchFn(ctx, () => {
			throwNotModel(node);
			if (node[LL_PREV] === null && node[LL_NEXT] === null && STATE.tail !== node) return false;
			removeLL(STATE, node);
			STATE.changes.push({
				kind: "remove",
				node
			});
			return true;
		});
	}, `${name}.remove`);
	const swap = (0, __reatom_core.action)((ctx, a, b) => {
		return batchFn(ctx, () => {
			throwNotModel(a);
			throwNotModel(b);
			if (a === b) return;
			swapLL(STATE, a, b);
			STATE.changes.push({
				kind: "swap",
				a,
				b
			});
		});
	}, `${name}.swap`);
	const move = (0, __reatom_core.action)((ctx, node, after) => {
		return batchFn(ctx, () => {
			throwNotModel(node);
			moveLL(STATE, node, after);
			STATE.changes.push({
				kind: "move",
				node,
				after
			});
		});
	}, `${name}.move`);
	const clear = (0, __reatom_core.action)((ctx) => {
		return batchFn(ctx, () => {
			clearLL(STATE);
			STATE.changes.push({ kind: "clear" });
		});
	}, `${name}.clear`);
	const find = (ctx, cb) => {
		for (let { head } = ctx.get(linkedList); head; head = head[LL_NEXT]) if (cb(head)) return head;
		return null;
	};
	const array = (0, __reatom_core.atom)((ctx, state = []) => toArray(ctx.spy(linkedList).head, state), `${name}.array`);
	const map = key ? (0, __reatom_core.atom)((ctx) => new Map(
		// use array as it already memoized and simplifies the order tracking
		ctx.spy(array).map((node) => {
			const keyValue = node[key];
			return [(0, __reatom_core.isAtom)(keyValue) ? ctx.spy(keyValue) : keyValue, node];
		})
)) : undefined;
	const reatomMap$1 = (cb, options$1 = {}) => {
		const { name: name$1 = (0, __reatom_core.__count)(`${_name}.reatomMap`),...hooks } = typeof options$1 === "string" ? { name: options$1 } : options$1;
		const mapList = (0, __reatom_core.atom)((ctx, mapList$1) => {
			(0, __reatom_core.throwReatomError)(STATE, `Can't compute the map of the linked list inside the batching.`);
			const ll = ctx.spy(linkedList);
			if (!mapList$1 || ll.version - 1 > mapList$1.version) {
				if (mapList$1) hooks.onClear?.(ctx, mapList$1);
				mapList$1 = {
					size: ll.size,
					version: ll.version,
					changes: [],
					head: null,
					tail: null,
					map: new WeakMap()
				};
				for (let head = ll.head; head; head = head[LL_NEXT]) {
					const node = cb(ctx, head);
					addLL(mapList$1, node, mapList$1.tail);
					mapList$1.map.set(head, node);
					hooks.onCreate?.(ctx, node);
				}
				mapList$1.size = ll.size;
			} else {
				mapList$1 = {
					head: mapList$1.head,
					tail: mapList$1.tail,
					size: mapList$1.size,
					version: ll.version,
					changes: [],
					map: mapList$1.map
				};
				for (const change of ll.changes) switch (change.kind) {
					case "create": {
						const node = cb(ctx, change.node);
						addLL(mapList$1, node, mapList$1.tail);
						mapList$1.map.set(change.node, node);
						mapList$1.changes.push({
							kind: "create",
							node
						});
						hooks.onCreate?.(ctx, node);
						break;
					}
					case "remove": {
						const node = mapList$1.map.get(change.node);
						removeLL(mapList$1, node);
						mapList$1.map.delete(change.node);
						mapList$1.changes.push({
							kind: "remove",
							node
						});
						hooks.onRemove?.(ctx, node, change.node);
						break;
					}
					case "swap": {
						const a = mapList$1.map.get(change.a);
						const b = mapList$1.map.get(change.b);
						swapLL(mapList$1, a, b);
						mapList$1.changes.push({
							kind: "swap",
							a,
							b
						});
						hooks.onSwap?.(ctx, {
							a,
							b
						});
						break;
					}
					case "move": {
						const node = mapList$1.map.get(change.node);
						const after = change.after ? mapList$1.map.get(change.after) : null;
						moveLL(mapList$1, node, after);
						mapList$1.changes.push({
							kind: "move",
							node,
							after
						});
						hooks.onMove?.(ctx, node);
						break;
					}
					case "clear": {
						hooks.onClear?.(ctx, mapList$1);
						clearLL(mapList$1);
						mapList$1.changes.push({ kind: "clear" });
						break;
					}
					default: {
						const kind = change;
						const error = new Error(`Unhandled linked list change "${kind}"`);
						throw error;
					}
				}
			}
			(0, __reatom_core.throwReatomError)(mapList$1.size !== ll.size, "Inconsistent linked list, is's a bug, please report an issue");
			return mapList$1;
		}, name$1);
		const array$1 = (0, __reatom_core.atom)((ctx, state = []) => toArray(ctx.spy(mapList).head, state), `${name$1}.array`);
		return Object.assign(mapList, {
			array: array$1,
			__reatomLinkedList: true
		});
	};
	return Object.assign(linkedList, {
		batch,
		create,
		remove,
		swap,
		move,
		clear,
		find,
		array,
		map,
		initiate: createLinkedList,
		reatomMap: reatomMap$1,
		__reatomLinkedList: true
	}).pipe(readonly);
}
const isLinkedListAtom = (thing) => thing?.__reatomLinkedList === true;

//#endregion
//#region src/reatomNumber.ts
const reatomNumber = (initState = 0, name) => (0, __reatom_core.atom)(initState, name).pipe(withAssign((target, name$1) => ({
	increment: (0, __reatom_core.action)((ctx, by = 1) => target(ctx, (prev) => prev + by), `${name$1}.increment`),
	decrement: (0, __reatom_core.action)((ctx, by = 1) => target(ctx, (prev) => prev - by), `${name$1}.decrement`),
	random: (0, __reatom_core.action)((ctx) => target(ctx, Math.random()), `${name$1}.decrement`),
	reset: (0, __reatom_core.action)((ctx) => target(ctx, initState), `${name$1}.reset`)
})));

//#endregion
//#region src/reatomRecord.ts
const reatomRecord = (initState, name) => (0, __reatom_core.atom)(initState, name).pipe(withAssign((target) => ({
	merge: (0, __reatom_core.action)((ctx, slice) => target(ctx, (prev) => {
		for (const key in prev) if (!Object.is(prev[key], slice[key])) return {
			...prev,
			...slice
		};
		return prev;
	}), `${name}.merge`),
	omit: (0, __reatom_core.action)((ctx, ...keys) => target(ctx, (prev) => {
		if (keys.some((key) => key in prev)) return (0, __reatom_utils.omit)(prev, keys);
		return prev;
	}), `${name}.omit`),
	reset: (0, __reatom_core.action)((ctx, ...keys) => target(ctx, (prev) => {
		if (keys.length === 0) return initState;
		const next = {};
		let changed = false;
		for (const key in prev) if (keys.includes(key)) if (key in initState) {
			next[key] = initState[key];
			changed ||= !Object.is(prev[key], initState[key]);
		} else changed ||= key in prev;
		else next[key] = prev[key];
		return changed ? next : prev;
	}), `${name}.reset`)
})));

//#endregion
//#region src/reatomSet.ts
const reatomSet = (initState = new Set(), name) => {
	const atomInitState = initState instanceof Set ? initState : new Set(initState);
	return (0, __reatom_core.atom)(atomInitState, name).pipe(withAssign((target, name$1) => ({
		__reatomSet: true,
		add: (0, __reatom_core.action)((ctx, el) => target(ctx, (prev) => prev.has(el) ? prev : new Set(prev).add(el)), `${name$1}.add`),
		set: (0, __reatom_core.action)((ctx, el) => target(ctx, (prev) => prev.has(el) ? prev : new Set(prev).add(el)), `${name$1}.set`),
		delete: (0, __reatom_core.action)((ctx, el) => {
			return target(ctx, (prev) => {
				if (!prev.has(el)) return prev;
				const next = new Set(prev);
				next.delete(el);
				return next;
			});
		}, `${name$1}.delete`),
		clear: (0, __reatom_core.action)((ctx) => {
			return target(ctx, (prev) => {
				if (prev.size === 0) return prev;
				return new Set();
			});
		}, `${name$1}.clear`),
		reset: (0, __reatom_core.action)((ctx) => target(ctx, atomInitState), `${name$1}.reset`),
		intersection: (0, __reatom_core.action)((ctx, set) => target(ctx, (prev) => prev.intersection(set)), `${name$1}.intersection`),
		union: (0, __reatom_core.action)((ctx, set) => target(ctx, (prev) => prev.union(set)), `${name$1}.union`),
		difference: (0, __reatom_core.action)((ctx, set) => target(ctx, (prev) => prev.difference(set)), `${name$1}.difference`),
		symmetricDifference: (0, __reatom_core.action)((ctx, set) => target(ctx, (prev) => prev.symmetricDifference(set)), `${name$1}.symmetricDifference`),
		toggle: (0, __reatom_core.action)((ctx, el) => {
			return target(ctx, (prev) => {
				if (!prev.has(el)) return new Set(prev).add(el);
				const next = new Set(prev);
				next.delete(el);
				return next;
			});
		}, `${name$1}.toggle`),
		has: (ctx, el) => ctx.get(target).has(el),
		isSubsetOf: (ctx, set) => ctx.get(target).isSubsetOf(set),
		isSupersetOf: (ctx, set) => ctx.get(target).isSupersetOf(set),
		isDisjointFrom: (ctx, set) => ctx.get(target).isDisjointFrom(set),
		size: (ctx) => ctx.get(target).size,
		sizeAtom: (0, __reatom_core.atom)((ctx) => ctx.spy(target).size, `${name$1}.size`)
	})));
};
const isSetAtom = (thing) => thing?.__reatomSet === true;

//#endregion
//#region src/reatomString.ts
const reatomString = (init = "", name) => (0, __reatom_core.atom)(init, name).pipe(withAssign((target, name$1) => ({ reset: (0, __reatom_core.action)((ctx) => target(ctx, init), `${name$1}.reset`) })));

//#endregion
//#region src/withComputed.ts
const withComputed = (computed) => (anAtom) => {
	const prevComputed = anAtom.__reatom.computer;
	anAtom.__reatom.computer = (ctx, state) => {
		if (prevComputed) state = prevComputed(ctx, state);
		return computed(ctx, state);
	};
	return anAtom;
};

//#endregion
//#region src/withReducers.ts
const withReducers = (reducers) => (anAtom) => Object.keys(reducers).reduce((anAtom$1, k) => {
	anAtom$1[k] = (0, __reatom_core.action)((ctx, ...args) => anAtom$1(ctx, reducers[k](ctx.get(anAtom$1), ...args)), `${anAtom$1.__reatom.name}._${k}`);
	return anAtom$1;
}, anAtom);

//#endregion
exports.LL_NEXT = LL_NEXT
exports.LL_PREV = LL_PREV
exports.isArrayAtom = isArrayAtom
exports.isLinkedListAtom = isLinkedListAtom
exports.isSetAtom = isSetAtom
exports.reatomArray = reatomArray
exports.reatomBoolean = reatomBoolean
exports.reatomEnum = reatomEnum
exports.reatomLinkedList = reatomLinkedList
exports.reatomMap = reatomMap
exports.reatomNumber = reatomNumber
exports.reatomRecord = reatomRecord
exports.reatomSet = reatomSet
exports.reatomString = reatomString
exports.withAssign = withAssign
exports.withComputed = withComputed
exports.withReducers = withReducers
