// @consta/table/__internal__/src/components/Table/TableRow/TableRow.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["className", "index", "lowHeadersAtom", "zebraStriped", "row", "tableRef", "leftNoVisibleItemsAtom", "rightNoVisibleItemsAtom"];
import "./TableRow.css";
import { useAtom } from '@reatom/npm-react';
import React, { forwardRef } from 'react';
import { cn } from "../../../utils/bem";
import { TableRowCell } from "../TableRowCell";
var cnTableRow = cn('TableRow');

var getMiss = function getMiss(colSpan, index, length, pinned) {
  if (pinned) {
    return 0;
  }

  if (colSpan === 'end') {
    return length - index - 1;
  }

  if (typeof colSpan === 'number') {
    return colSpan - 1;
  }

  return 0;
};

export var TableRow = forwardRef(function (props, ref) {
  var className = props.className,
      rowIndex = props.index,
      lowHeadersAtom = props.lowHeadersAtom,
      zebraStriped = props.zebraStriped,
      row = props.row,
      tableRef = props.tableRef,
      leftNoVisibleItemsAtom = props.leftNoVisibleItemsAtom,
      rightNoVisibleItemsAtom = props.rightNoVisibleItemsAtom,
      otherProps = _objectWithoutProperties(props, _excluded);

  var _useAtom = useAtom(lowHeadersAtom),
      _useAtom2 = _slicedToArray(_useAtom, 1),
      lowHeaders = _useAtom2[0];

  var _useAtom3 = useAtom(leftNoVisibleItemsAtom),
      _useAtom4 = _slicedToArray(_useAtom3, 1),
      leftNoVisibleItems = _useAtom4[0];

  var _useAtom5 = useAtom(rightNoVisibleItemsAtom),
      _useAtom6 = _slicedToArray(_useAtom5, 1),
      rightNoVisibleItems = _useAtom6[0];

  var miss = 0;
  var nodes = [];

  for (var columnIndex = 0; columnIndex < lowHeaders.length; columnIndex++) {
    var _lowHeaders, _lowHeaders2;

    var _lowHeaders$columnInd = lowHeaders[columnIndex],
        isSeparator = _lowHeaders$columnInd.isSeparator,
        accessor = _lowHeaders$columnInd.accessor,
        pinned = _lowHeaders$columnInd.pinned,
        RenderCell = _lowHeaders$columnInd.renderCell,
        getColSpan = _lowHeaders$columnInd.colSpan;

    if (miss) {
      miss = miss ? miss - 1 : miss;
      continue;
    }

    miss = getMiss(getColSpan === null || getColSpan === void 0 ? void 0 : getColSpan({
      row: row
    }), columnIndex, lowHeaders.length, !!pinned);

    if (!pinned && columnIndex !== 0 && leftNoVisibleItems >= 1 && leftNoVisibleItems - columnIndex - miss >= 1 || !pinned && columnIndex !== 0 && rightNoVisibleItems >= 0 && rightNoVisibleItems + columnIndex >= lowHeaders.length) {
      continue;
    }

    nodes.push(React.createElement(TableRowCell, {
      key: cnTableRow('Cell', {
        key: accessor || columnIndex
      }),
      borderLeft: columnIndex !== 0 && !(pinned !== 'left' && ((_lowHeaders = lowHeaders[columnIndex - 1]) === null || _lowHeaders === void 0 ? void 0 : _lowHeaders.pinned) === 'left'),
      borderRight: pinned === 'left' && ((_lowHeaders2 = lowHeaders[columnIndex + 1]) === null || _lowHeaders2 === void 0 ? void 0 : _lowHeaders2.pinned) !== 'left',
      borderTop: !isSeparator && rowIndex !== 0,
      ref: columnIndex === 0 ? ref : undefined,
      row: row,
      rowIndex: rowIndex,
      miss: miss,
      index: columnIndex,
      tableRef: tableRef,
      accessor: accessor,
      separator: isSeparator,
      pinned: pinned,
      renderCell: RenderCell
    }));
  }

  return React.createElement("div", Object.assign({}, otherProps, {
    className: cnTableRow({
      zebraStriped: zebraStriped
    }, [className]),
    "aria-hidden": "true"
  }), nodes);
});
