// @consta/table/__internal__/src/components/Toolbar/Toolbar.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["leftSide", "rightSide", "className", "form", "border", "space", "itemsGap"];
import "./Toolbar.css";
import { isNotNil } from '@consta/uikit/__internal__/src/utils/type-guards';
import { cnMixFlex } from '@consta/uikit/MixFlex';
import { cnMixSpace } from '@consta/uikit/MixSpace';
import React, { forwardRef } from 'react';
import { cn } from "../../utils/bem";
var cnToolbar = cn('Toolbar');

var renderSlot = function renderSlot(slot) {
  if (Array.isArray(slot)) {
    return slot.filter(isNotNil).map(function (item, index) {
      return React.createElement("div", {
        className: cnToolbar('Slot'),
        key: index
      }, item);
    });
  }

  return renderSlot([slot]);
};

var guardBorder = function guardBorder(border, form) {
  if (border === 'top' && form === 'defaultBrick' || border === 'bottom' && form === 'brickDefault' || border === 'top' && form === 'default' || border === 'bottom' && form === 'default') {
    return undefined;
  }

  if (border === true) {
    return 'all';
  }

  return border;
};

var getGap = function getGap(gap, index) {
  if (Array.isArray(gap)) {
    return gap[index];
  }

  return gap;
};

export var Toolbar = forwardRef(function (_ref, ref) {
  var leftSide = _ref.leftSide,
      rightSide = _ref.rightSide,
      className = _ref.className,
      _ref$form = _ref.form,
      form = _ref$form === void 0 ? 'default' : _ref$form,
      border = _ref.border,
      _ref$space = _ref.space,
      space = _ref$space === void 0 ? {
    pH: 's',
    pV: 'xs'
  } : _ref$space,
      _ref$itemsGap = _ref.itemsGap,
      itemsGap = _ref$itemsGap === void 0 ? 's' : _ref$itemsGap,
      otherProps = _objectWithoutProperties(_ref, _excluded);

  return React.createElement("div", Object.assign({}, otherProps, {
    ref: ref,
    className: cnToolbar({
      form: form,
      border: guardBorder(border, form)
    }, [cnMixFlex({
      justify: rightSide && !leftSide ? 'flex-end' : 'space-between',
      gap: 'm'
    }), cnMixSpace(space), className])
  }), leftSide && React.createElement("div", {
    className: cnMixFlex({
      gap: getGap(itemsGap, 0)
    })
  }, renderSlot(leftSide)), rightSide && React.createElement("div", {
    className: cnMixFlex({
      gap: getGap(itemsGap, 1),
      justify: 'flex-end'
    })
  }, renderSlot(rightSide)));
});
