// @consta/table/__internal__/src/components/Pagination/Pagination.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["label", "step", "steps", "total", "offset", "offsetLabel", "className", "buttonPrevRef", "buttonNextRef", "onStepChange", "onChange"];
import "./Pagination.css";
import { IconArrowLeft } from '@consta/icons/IconArrowLeft';
import { IconArrowRight } from '@consta/icons/IconArrowRight';
import { withCtx } from '@consta/uikit/__internal__/src/utils/state/withCtx';
import { Button } from '@consta/uikit/Button';
import { cnMixFlex } from '@consta/uikit/MixFlex';
import { Select } from '@consta/uikit/SelectCanary';
import { Text } from '@consta/uikit/Text';
import { useMutableRef } from '@consta/uikit/useMutableRef';
import { useAction, useAtom, useUpdate } from '@reatom/npm-react';
import React, { forwardRef } from 'react';
import { cn } from "../../utils/bem";
import { isFunction, isNumber } from "../../utils/type-guards";
var cnPagination = cn('Pagination');

var getItem = function getItem(item) {
  return item.toString();
};

var defaultOffsetLabel = function defaultOffsetLabel(offset, step, total) {
  var from = offset + 1;
  var to = offset + step;

  if (isNumber(total)) {
    return "".concat(from, "-").concat(to > total ? total : to, " \u0438\u0437 ").concat(total);
  }

  return "".concat(from, "-").concat(to);
};

var guardOffset = function guardOffset(value, total) {
  if (isNumber(total)) {
    return Math.max(Math.min(value, total), 0);
  }

  return Math.max(value, 0);
};

export var Pagination = withCtx(forwardRef(function (props, ref) {
  var _props$label = props.label,
      label = _props$label === void 0 ? 'Строк на странице' : _props$label,
      _props$step = props.step,
      step = _props$step === void 0 ? 10 : _props$step,
      _props$steps = props.steps,
      steps = _props$steps === void 0 ? [10, 25, 50, 100] : _props$steps,
      total = props.total,
      _props$offset = props.offset,
      offset = _props$offset === void 0 ? 0 : _props$offset,
      _props$offsetLabel = props.offsetLabel,
      offsetLabel = _props$offsetLabel === void 0 ? defaultOffsetLabel : _props$offsetLabel,
      className = props.className,
      buttonPrevRef = props.buttonPrevRef,
      buttonNextRef = props.buttonNextRef,
      onStepChange = props.onStepChange,
      onChange = props.onChange,
      otherProps = _objectWithoutProperties(props, _excluded);

  var refs = useMutableRef([onStepChange, total, onChange]);

  var _useAtom = useAtom(step),
      _useAtom2 = _slicedToArray(_useAtom, 3),
      stepValue = _useAtom2[0],
      setStepValue = _useAtom2[1],
      stepValueAtom = _useAtom2[2];

  var _useAtom3 = useAtom(guardOffset(offset, total)),
      _useAtom4 = _slicedToArray(_useAtom3, 3),
      offsetValue = _useAtom4[0],
      setOffsetValue = _useAtom4[1],
      offsetValueAtom = _useAtom4[2];

  var onSelectChange = useAction(function (ctx, value, props) {
    if (isNumber(value)) {
      var _refs$current$, _refs$current;

      stepValueAtom(ctx, value);
      (_refs$current$ = (_refs$current = refs.current)[0]) === null || _refs$current$ === void 0 ? void 0 : _refs$current$.call(_refs$current, value, props);
    }
  });
  var nextPageFactory = useAction(function (ctx, e) {
    var _refs$current$2, _refs$current2;

    var newValue = guardOffset(ctx.get(offsetValueAtom) + ctx.get(stepValueAtom), refs.current[1]);
    offsetValueAtom(ctx, newValue);
    (_refs$current$2 = (_refs$current2 = refs.current)[2]) === null || _refs$current$2 === void 0 ? void 0 : _refs$current$2.call(_refs$current2, newValue, {
      e: e
    });
  });
  var prevPageFactory = useAction(function (ctx, e) {
    var _refs$current$3, _refs$current3;

    var newValue = guardOffset(ctx.get(offsetValueAtom) - ctx.get(stepValueAtom), refs.current[1]);
    offsetValueAtom(ctx, newValue);
    (_refs$current$3 = (_refs$current3 = refs.current)[2]) === null || _refs$current$3 === void 0 ? void 0 : _refs$current$3.call(_refs$current3, newValue, {
      e: e
    });
  });
  useUpdate(stepValueAtom, [step]);
  useUpdate(offsetValueAtom, [offset]);
  return React.createElement("div", Object.assign({
    ref: ref
  }, otherProps, {
    className: cnPagination(null, [cnMixFlex({
      align: 'center',
      gap: 'xl'
    }), className])
  }), React.createElement("div", {
    className: cnMixFlex({
      align: 'center',
      gap: 's'
    })
  }, label && React.createElement(Text, {
    size: "s",
    view: "secondary"
  }, label), steps && React.createElement(Select, {
    className: cnPagination('Select'),
    items: steps,
    getItemLabel: getItem,
    getItemKey: getItem,
    onChange: onSelectChange,
    value: stepValue,
    size: "s"
  })), React.createElement("div", {
    className: cnMixFlex({
      align: 'center',
      gap: 's'
    })
  }, React.createElement(Text, {
    size: "s",
    className: cnPagination('OffsetLabel')
  }, isFunction(offsetLabel) ? offsetLabel(offsetValue, stepValue, total) : offsetLabel), React.createElement(Button, {
    ref: buttonPrevRef,
    iconLeft: IconArrowLeft,
    onlyIcon: true,
    view: "clear",
    size: "s",
    disabled: offsetValue <= 0,
    onClick: prevPageFactory
  }), React.createElement(Button, {
    ref: buttonNextRef,
    iconLeft: IconArrowRight,
    onlyIcon: true,
    view: "clear",
    size: "s",
    onClick: nextPageFactory,
    disabled: total ? offsetValue >= total - stepValue : undefined
  })));
}));
