// @consta/table/__internal__/src/components/DataCell/DataCell.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["className", "control", "icon", "children", "level", "view", "size", "indicator", "style", "truncate"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import "./DataCell.css";
import { cnMixFlex } from '@consta/uikit/MixFlex';
import { cnMixSpace } from '@consta/uikit/MixSpace';
import { Text } from '@consta/uikit/Text';
import React, { forwardRef, Fragment } from 'react';
import { cn } from "../../utils/bem";
import { cellVerticalSpaceMap } from "../../utils/maps/cellVerticalSpaceMap";
import { isNumber, isString } from "../../utils/type-guards";
var cnDataCell = cn('DataCell');

var renderContentSlot = function renderContentSlot(children, truncate) {
  return React.createElement("div", {
    className: cnDataCell('ContentSlot', {
      truncate: truncate
    }, [cnMixFlex({
      flex: 'flex',
      align: 'center'
    })])
  }, children);
};

var renderIcon = function renderIcon(Icon, view) {
  return renderContentSlot(React.createElement(Icon, {
    view: view,
    size: "s"
  }));
};

var renderChildren = function renderChildren(children, view, size, truncate) {
  if (isString(children) || isNumber(children)) {
    return renderContentSlot(React.createElement(Text, {
      className: cnMixSpace({
        pV: cellVerticalSpaceMap[size]
      }),
      view: view,
      size: size,
      truncate: truncate
    }, children), truncate);
  }

  return renderContentSlot(children);
};

export var DataCell = forwardRef(function (props, ref) {
  var _objectSpread2;

  var className = props.className,
      control = props.control,
      icon = props.icon,
      children = props.children,
      _props$level = props.level,
      levelProp = _props$level === void 0 ? 0 : _props$level,
      view = props.view,
      _props$size = props.size,
      size = _props$size === void 0 ? 'm' : _props$size,
      indicator = props.indicator,
      style = props.style,
      _props$truncate = props.truncate,
      truncate = _props$truncate === void 0 ? false : _props$truncate,
      otherProps = _objectWithoutProperties(props, _excluded);

  var level = levelProp < 0 ? 0 : levelProp;
  var controlsSlots = control ? _toConsumableArray(Array.isArray(control) ? control : [control]) : [];
  var iconsSlots = icon ? _toConsumableArray(Array.isArray(icon) ? icon.map(function (icon) {
    return renderIcon(icon, view);
  }) : [renderIcon(icon, view)]) : [];
  var childrenSlots = children ? _toConsumableArray(Array.isArray(children) ? children.map(function (item) {
    return renderChildren(item, view, size, truncate);
  }) : [renderChildren(children, view, size, truncate)]) : [];
  var contentSlots = [].concat(_toConsumableArray(iconsSlots), _toConsumableArray(childrenSlots));
  return React.createElement("div", Object.assign({}, otherProps, {
    className: cnDataCell({
      size: size,
      indicator: Boolean(indicator),
      alignmentIndent: level >= 1 && controlsSlots.length === 0
    }, [cnMixFlex({
      flex: 'flex'
    }), className]),
    style: _objectSpread(_objectSpread({}, style), {}, (_objectSpread2 = {}, _defineProperty(_objectSpread2, '--table-data-cell-level', level || undefined), _defineProperty(_objectSpread2, '--table-data-cell-indicator-color', indicator ? "var(--color-bg-".concat(indicator, ")") : undefined), _objectSpread2)),
    ref: ref
  }), controlsSlots.length ? React.createElement("div", {
    className: cnMixFlex({
      flex: 'flex',
      align: 'center',
      gap: '2xs'
    })
  }, controlsSlots.map(function (item, index) {
    return React.createElement("div", {
      className: cnDataCell('ControlSlot', [cnMixFlex({
        flex: 'flex',
        align: 'center',
        justify: 'center'
      })]),
      key: index
    }, item);
  })) : undefined, contentSlots.length ? React.createElement("div", {
    className: cnDataCell('Slots', {
      truncate: truncate
    }, [cnMixFlex({
      flex: 'flex',
      gap: '2xs'
    })])
  }, contentSlots.map(function (item, index) {
    return React.createElement(Fragment, {
      key: index
    }, item);
  })) : undefined);
});
