// @consta/table/__internal__/src/components/Table/TableData/TableData.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["className", "rows", "lowHeadersAtom", "rowsRefsAtom", "sliceAtom", "zebraStriped", "onRowMouseEnter", "onRowMouseLeave", "onRowClick", "getRowKey", "tableRef", "rowHoverEffect", "leftNoVisibleItemsAtom", "rightNoVisibleItemsAtom"];
import "./TableData.css";
import { useAtom } from '@reatom/npm-react';
import React, { forwardRef } from 'react';
import { cn } from "../../../utils/bem";
import { isNotNil } from "../../../utils/type-guards";
import { TableRow } from "../TableRow";
export var cnTableData = cn('TableData');

var getRowMouseEvent = function getRowMouseEvent(row, fn) {
  return fn ? function (e) {
    return fn(row, {
      e: e
    });
  } : undefined;
};

var getKey = function getKey(row, getRowKey, rowIndex) {
  var key = getRowKey === null || getRowKey === void 0 ? void 0 : getRowKey(row);

  if (isNotNil(key)) {
    return key;
  }

  if (typeof row.id === 'string' || typeof row.id === 'number') {
    return row.id;
  }

  return rowIndex;
};

var TableDataRender = function TableDataRender(props, ref) {
  var className = props.className,
      rows = props.rows,
      lowHeadersAtom = props.lowHeadersAtom,
      rowsRefsAtom = props.rowsRefsAtom,
      sliceAtom = props.sliceAtom,
      _props$zebraStriped = props.zebraStriped,
      zebraStriped = _props$zebraStriped === void 0 ? false : _props$zebraStriped,
      onRowMouseEnter = props.onRowMouseEnter,
      onRowMouseLeave = props.onRowMouseLeave,
      onRowClick = props.onRowClick,
      getRowKey = props.getRowKey,
      tableRef = props.tableRef,
      rowHoverEffect = props.rowHoverEffect,
      leftNoVisibleItemsAtom = props.leftNoVisibleItemsAtom,
      rightNoVisibleItemsAtom = props.rightNoVisibleItemsAtom,
      otherProps = _objectWithoutProperties(props, _excluded);

  var _useAtom = useAtom(rowsRefsAtom),
      _useAtom2 = _slicedToArray(_useAtom, 1),
      rowsRefs = _useAtom2[0];

  var _useAtom3 = useAtom(sliceAtom),
      _useAtom4 = _slicedToArray(_useAtom3, 1),
      slice = _useAtom4[0];

  return React.createElement("div", Object.assign({}, otherProps, {
    ref: ref,
    className: cnTableData({
      rowHoverEffect: rowHoverEffect
    }, [className])
  }), rows.slice.apply(rows, _toConsumableArray(slice)).map(function (row, index) {
    var rowIndex = index + slice[0];
    var rowZebraStriped = zebraStriped && rowIndex % 2 !== 0;
    return React.createElement(TableRow, {
      key: getKey(row, getRowKey, rowIndex),
      row: row,
      index: rowIndex,
      lowHeadersAtom: lowHeadersAtom,
      zebraStriped: rowZebraStriped,
      onMouseEnter: getRowMouseEvent(row, onRowMouseEnter),
      onMouseLeave: getRowMouseEvent(row, onRowMouseLeave),
      onClick: getRowMouseEvent(row, onRowClick),
      tableRef: tableRef,
      ref: rowsRefs[rowIndex],
      leftNoVisibleItemsAtom: leftNoVisibleItemsAtom,
      rightNoVisibleItemsAtom: rightNoVisibleItemsAtom
    });
  }));
};

export var TableData = forwardRef(TableDataRender);
