// @consta/table/__internal__/src/components/Table/TableHeader/TableHeader.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["headersAtom", "className", "stickyLeftOffsetsAtom", "stickyRightOffsetsAtom", "stickyHeader", "headerCellsRefsAtom", "bordersAtom", "tableRef"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import "./TableHeader.css";
import { forkRef } from '@consta/uikit/useForkRef';
import { useAtom } from '@reatom/npm-react';
import React, { forwardRef } from 'react';
import { HeaderDataCell } from "../../HeaderDataCell";
import { cn } from "../../../utils/bem";
import { cnTableCell } from "../TableCell";
export var cnTableHeader = cn('TableHeader');

var DefaultRenderHeaderCell = function DefaultRenderHeaderCell(_ref) {
  var title = _ref.title;
  return React.createElement(HeaderDataCell, null, title);
};

export var TableHeader = forwardRef(function (props, ref) {
  var headersAtom = props.headersAtom,
      className = props.className,
      stickyLeftOffsetsAtom = props.stickyLeftOffsetsAtom,
      stickyRightOffsetsAtom = props.stickyRightOffsetsAtom,
      stickyHeader = props.stickyHeader,
      headerCellsRefsAtom = props.headerCellsRefsAtom,
      bordersAtom = props.bordersAtom,
      tableRef = props.tableRef,
      otherProps = _objectWithoutProperties(props, _excluded);

  var _useAtom = useAtom(headersAtom),
      _useAtom2 = _slicedToArray(_useAtom, 1),
      headers = _useAtom2[0];

  var _useAtom3 = useAtom(bordersAtom),
      _useAtom4 = _slicedToArray(_useAtom3, 1),
      borders = _useAtom4[0];

  var _useAtom5 = useAtom(stickyLeftOffsetsAtom),
      _useAtom6 = _slicedToArray(_useAtom5, 1),
      stickyLeftOffsets = _useAtom6[0];

  var _useAtom7 = useAtom(stickyRightOffsetsAtom),
      _useAtom8 = _slicedToArray(_useAtom7, 1),
      stickyRightOffsets = _useAtom8[0];

  var _useAtom9 = useAtom(headerCellsRefsAtom),
      _useAtom10 = _slicedToArray(_useAtom9, 1),
      headerCellsRefs = _useAtom10[0];

  return React.createElement("div", Object.assign({}, otherProps, {
    className: cnTableHeader(null, [className]),
    ref: ref
  }), headers.map(function (column, index) {
    var style = {};

    if (column.position.colSpan) {
      style.gridColumnEnd = "span ".concat(column.position.colSpan);
    }

    if (column.position.rowSpan) {
      style.gridRowEnd = "span ".concat(column.position.rowSpan);
    }

    var RenderHeaderCell = column.renderHeaderCell || DefaultRenderHeaderCell;
    return React.createElement("div", {
      className: cnTableHeader('Cell', {
        pinned: !!column.pinned
      }, [cnTableCell({
        separator: column.isSeparator,
        borderLeft: borders[index][0],
        borderRight: borders[index][1],
        borderTop: borders[index][2],
        sticky: !!column.pinned || stickyHeader,
        up: !!column.pinned
      })]),
      style: _objectSpread(_objectSpread({}, style), {}, {
        top: stickyHeader ? "var(--table-column-sticky-top-offset-".concat(index, ")") : undefined,
        left: column.pinned === 'left' ? "var(--table-column-sticky-left-offset-".concat(stickyLeftOffsets[index], ")") : undefined,
        right: column.pinned === 'right' ? "var(--table-column-sticky-right-offset-".concat(stickyRightOffsets[index], ")") : undefined
      }),
      ref: forkRef([headerCellsRefs[index]]),
      key: cnTableHeader('Cell', {
        index: index
      })
    }, column.isSeparator ? null : React.createElement(RenderHeaderCell, {
      title: column.title,
      index: index,
      tableRef: tableRef
    }));
  }));
});
