// @consta/table/__internal__/src/components/Table/TableBody/helpers.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import { range } from "../../../utils/array";
import { isNumber, isString } from "../../../utils/type-guards";
export var getStyleTopOffsets = function getStyleTopOffsets(topOffsets) {
  var result = {};

  for (var _index = 0; _index < topOffsets.length; _index++) {
    result["--table-lowheader-top-offset-".concat(_index)] = "".concat(topOffsets[_index], "px");
  }

  return result;
};
export var getStyleByArray = function getStyleByArray(array, cssVarName) {
  var print = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'px';
  var result = '';

  for (var _index2 = 0; _index2 < array.length; _index2++) {
    result += "".concat(cssVarName, "-").concat(_index2, ": ").concat(isString(print) ? array[_index2] + print : print(array[_index2], _index2, array), "; ");
  }

  return result;
};
export var getStyleLeftOffsetsForStickyColumns = function getStyleLeftOffsetsForStickyColumns(length) {
  return getStyleByArray(range(length), '--table-column-sticky-left-offset', printTableColumnStickyHorizontalOffsetValue());
};
export var getStyleRightOffsetsForStickyColumns = function getStyleRightOffsetsForStickyColumns(length) {
  return getStyleByArray(range(length), '--table-column-sticky-right-offset', printTableColumnStickyHorizontalOffsetValue(true));
};
export var getGridTemplate = function getGridTemplate(length) {
  return _toConsumableArray(Array(length)).map(function (_, index) {
    return "var(--table-column-size-".concat(index, ")");
  }).join(' ');
};

var printHorizontalOffsetPart = function printHorizontalOffsetPart(reverse, array) {
  return function (value) {
    return reverse ? "var(--table-column-size-".concat(array.length - 1 - value, ")") : "var(--table-column-size-".concat(value, ")");
  };
};

var printTableColumnStickyHorizontalOffsetValue = function printTableColumnStickyHorizontalOffsetValue() {
  var reverse = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
  return function (index, _, array) {
    var indexArray = range(reverse ? array.length - 1 - index : index);
    var value = indexArray.map(printHorizontalOffsetPart(reverse, array));

    if (!value.length) {
      return '0px';
    }

    if (value.length === 1) {
      return value[0];
    }

    return "calc(".concat(value.join(' + '), ")");
  };
};

export var printSize = function printSize(value) {
  if (isNumber(value)) {
    return "".concat(value, "px");
  }

  if (isString(value)) {
    return value;
  }

  return 'auto';
};
