// @consta/table/__internal__/src/components/HeaderDataCell/HeaderDataCell.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["className", "size", "controlRight", "children", "controlLeft"];
import "./HeaderDataCell.css";
import { cnMixFlex } from '@consta/uikit/MixFlex';
import { cnMixSpace } from '@consta/uikit/MixSpace';
import { Text } from '@consta/uikit/Text';
import React, { forwardRef, Fragment } from 'react';
import { cn } from "../../utils/bem";
import { cellVerticalSpaceMap } from "../../utils/maps/cellVerticalSpaceMap";
import { isNumber, isString } from "../../utils/type-guards";
var cnHeaderDataCell = cn('HeaderDataCell');

var renderContentSlot = function renderContentSlot(children) {
  return React.createElement("div", {
    className: cnHeaderDataCell('ContentSlot', [cnMixFlex({
      flex: 'flex',
      align: 'center'
    })])
  }, children);
};

var renderChildren = function renderChildren(children, size) {
  if (isString(children) || isNumber(children)) {
    return renderContentSlot(React.createElement(Text, {
      className: cnMixSpace({
        pV: cellVerticalSpaceMap[size]
      }),
      size: size
    }, children));
  }

  return renderContentSlot(children);
};

export var HeaderDataCell = forwardRef(function (props, ref) {
  var className = props.className,
      _props$size = props.size,
      size = _props$size === void 0 ? 'm' : _props$size,
      controlRight = props.controlRight,
      children = props.children,
      controlLeft = props.controlLeft,
      otherProps = _objectWithoutProperties(props, _excluded);

  var childrenSlots = children ? _toConsumableArray(Array.isArray(children) ? children.map(function (item) {
    return renderChildren(item, size);
  }) : [renderChildren(children, size)]) : [];
  var controlRightSlots = controlRight ? _toConsumableArray(Array.isArray(controlRight) ? controlRight : [controlRight]) : [];
  var controlLeftSlots = controlLeft ? _toConsumableArray(Array.isArray(controlLeft) ? controlLeft : [controlLeft]) : [];
  return React.createElement("div", Object.assign({}, otherProps, {
    className: cnHeaderDataCell({
      size: size
    }, [cnMixFlex({
      flex: 'flex',
      gap: 'xs',
      justify: 'space-between'
    }), cnMixSpace({
      pH: 's'
    }), className]),
    ref: ref
  }), React.createElement("div", {
    className: cnMixFlex({
      flex: 'flex',
      gap: 'xs'
    })
  }, controlLeftSlots.length ? React.createElement("div", {
    className: cnMixFlex({
      flex: 'flex',
      gap: '2xs'
    })
  }, controlLeftSlots.map(function (item, index) {
    return React.createElement("div", {
      className: cnHeaderDataCell('ControlSlot', [cnMixFlex({
        flex: 'flex',
        align: 'center',
        justify: 'center'
      })]),
      key: index
    }, item);
  })) : undefined, childrenSlots.length ? React.createElement("div", {
    className: cnMixFlex({
      flex: 'flex',
      gap: '2xs'
    })
  }, childrenSlots.map(function (item, index) {
    return React.createElement(Fragment, {
      key: index
    }, item);
  })) : undefined), controlRightSlots.length ? React.createElement("div", {
    className: cnMixFlex({
      flex: 'flex',
      align: 'center',
      gap: '2xs'
    })
  }, controlRightSlots.map(function (item, index) {
    return React.createElement("div", {
      className: cnHeaderDataCell('ControlSlot', [cnMixFlex({
        flex: 'flex',
        align: 'center',
        justify: 'center'
      })]),
      key: index
    }, item);
  })) : undefined);
});
