// @consta/table/__internal__/src/components/Table/TableResizers/TableResizers.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import "./TableResizers.css";
import { useSendToAtom } from '@consta/uikit/__internal__/src/utils/state/useSendToAtom';
import { useRefs } from '@consta/uikit/useRefs';
import { useAtom } from '@reatom/npm-react';
import React, { forwardRef, memo } from 'react';
import { cn } from "../../../utils/bem";
import { useVisibleColumns } from "./useVisibleColumns";
var cnTableResizers = cn('TableResizers');
var TableResizer = forwardRef(function (_ref, ref) {
  var resizableAtom = _ref.resizableAtom,
      maxWidth = _ref.maxWidth,
      minWidth = _ref.minWidth,
      isSeparator = _ref.isSeparator,
      pinned = _ref.pinned,
      index = _ref.index,
      lowHeadersLength = _ref.lowHeadersLength,
      handlersAtom = _ref.handlersAtom,
      activeIndexAtom = _ref.activeIndexAtom,
      virtualScrollHelperRef = _ref.virtualScrollHelperRef;

  var _useAtom = useAtom(resizableAtom),
      _useAtom2 = _slicedToArray(_useAtom, 1),
      resizable = _useAtom2[0];

  var _useAtom3 = useAtom(handlersAtom),
      _useAtom4 = _slicedToArray(_useAtom3, 1),
      handlers = _useAtom4[0];

  var indexAtom = useSendToAtom(index);

  var _useAtom5 = useAtom(function (ctx) {
    return ctx.spy(indexAtom) === ctx.spy(activeIndexAtom);
  }),
      _useAtom6 = _slicedToArray(_useAtom5, 1),
      active = _useAtom6[0];

  return React.createElement("div", {
    className: cnTableResizers('Cell'),
    ref: ref,
    style: _defineProperty({}, '--table-resizer-top-offset', "var(--table-resizer-top-offset-".concat(index, ")"))
  }, React.createElement("div", {
    ref: virtualScrollHelperRef,
    className: cnTableResizers('VirtualScrollHelper')
  }), resizable && (maxWidth === undefined || maxWidth !== minWidth) && !isSeparator && !pinned && !(resizable === 'inside' && lowHeadersLength === index + 1) && React.createElement("div", Object.assign({}, handlers[index], {
    className: cnTableResizers('Resizer', {
      active: active
    }),
    "aria-hidden": true
  })));
});
export var TableResizers = memo(function (props) {
  var lowHeadersAtom = props.lowHeadersAtom,
      resizersRefsAtom = props.resizersRefsAtom,
      handlersAtom = props.handlersAtom,
      resizableAtom = props.resizableAtom,
      activeIndexAtom = props.activeIndexAtom,
      intersectingColumnsAtom = props.intersectingColumnsAtom,
      bodyElAtom = props.bodyElAtom;

  var _useAtom7 = useAtom(resizersRefsAtom),
      _useAtom8 = _slicedToArray(_useAtom7, 1),
      resizersRefs = _useAtom8[0];

  var _useAtom9 = useAtom(lowHeadersAtom),
      _useAtom10 = _slicedToArray(_useAtom9, 1),
      lowHeaders = _useAtom10[0];

  var virtualScrollHelperRefs = useRefs(lowHeaders.length);
  useVisibleColumns(virtualScrollHelperRefs, intersectingColumnsAtom, bodyElAtom);
  return React.createElement("div", {
    className: cnTableResizers()
  }, lowHeaders.map(function (_ref3, index) {
    var maxWidth = _ref3.maxWidth,
        minWidth = _ref3.minWidth,
        pinned = _ref3.pinned,
        isSeparator = _ref3.isSeparator;
    return React.createElement(TableResizer, {
      resizableAtom: resizableAtom,
      key: index,
      ref: resizersRefs[index],
      virtualScrollHelperRef: virtualScrollHelperRefs[index],
      maxWidth: maxWidth,
      minWidth: minWidth,
      pinned: pinned,
      isSeparator: isSeparator,
      index: index,
      lowHeadersLength: lowHeaders.length,
      handlersAtom: handlersAtom,
      activeIndexAtom: activeIndexAtom
    });
  }));
});
