// @consta/table/__internal__/src/components/Collapse/CollapseFullscreen/CollapseFullscreen.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import "./CollapseFullscreen.css";
import { animateTimeout, cnMixPopoverAnimate } from '@consta/uikit/MixPopoverAnimate';
import { PortalWithTheme, PortalWithThemeConsumer } from '@consta/uikit/PortalWithTheme';
import { useTheme } from '@consta/uikit/Theme';
import { getElementHeight } from '@consta/uikit/useResizeObserved';
import { useAtom } from '@reatom/npm-react';
import React, { useEffect, useMemo, useRef } from 'react';
import { Transition } from 'react-transition-group';
import { Toolbar } from "../../Toolbar";
import { useResizeObservedAtom } from "../../../hooks/useResizeObservedAtom";
import { cn } from "../../../utils/bem";
import { CollapseButton } from "../CollapseButton";
import { CollapseFullscreenIcon } from "../CollapseFullscreenIcon";
var cnCollapseFullscreen = cn('CollapseFullscreen');

var isClassName = function isClassName(element, className) {
  return element.classList.contains(className);
};

var addClassName = function addClassName(element, className) {
  if (!isClassName(element, className)) {
    element.classList.add(className);
  }
};

var removeClassName = function removeClassName(element, className) {
  if (isClassName(element, className)) {
    element.classList.remove(className);
  }
};

var addCssVariable = function addCssVariable(variableName, value) {
  document.documentElement.style.setProperty(variableName, value);
};

var removeCssVariable = function removeCssVariable(variableName) {
  document.documentElement.style.removeProperty(variableName);
};

var getScrollBarWidth = function getScrollBarWidth() {
  return window.innerWidth - document.documentElement.clientWidth;
};

var bodyAddStyles = function bodyAddStyles() {
  var scrollBarWidth = getScrollBarWidth();
  addCssVariable('--ct-collapse-fullscreen-body-scrollbar-width', "".concat(scrollBarWidth, "px"));
  addClassName(window.document.documentElement, cnCollapseFullscreen('Body'));
};

var bodyRemoveStyles = function bodyRemoveStyles() {
  removeClassName(window.document.documentElement, cnCollapseFullscreen('Body'));
  removeCssVariable('--ct-collapse-fullscreen-body-scrollbar-width');
};

export var CollapseFullscreen = function CollapseFullscreen(_ref) {
  var _ref$active = _ref.active,
      active = _ref$active === void 0 ? false : _ref$active,
      container = _ref.container,
      zIndex = _ref.zIndex,
      rightSide = _ref.rightSide,
      onFullscreen = _ref.onFullscreen,
      children = _ref.children,
      leftSide = _ref.leftSide;
  var containerIsBody = container === window.document.body || 'current' in container && container.current === window.document.body;
  var portalRef = useRef(null);

  var _useTheme = useTheme(),
      theme = _useTheme.theme;

  var toolbarRef = useRef(null);
  var toolbarHeight = useAtom(useResizeObservedAtom(useMemo(function () {
    return [toolbarRef];
  }, [toolbarRef, active]), getElementHeight))[0][0];
  useEffect(function () {
    if (active && containerIsBody) {
      bodyAddStyles();
    }
  }, [active]);
  useEffect(function () {
    return bodyRemoveStyles;
  }, []);
  return React.createElement(Transition, {
    "in": active,
    unmountOnExit: true,
    nodeRef: portalRef,
    timeout: animateTimeout,
    onExited: bodyRemoveStyles
  }, function (animate) {
    return React.createElement(PortalWithTheme, {
      ref: portalRef,
      preset: theme,
      container: container,
      style: _defineProperty({
        zIndex: zIndex
      }, "--collapse-toolbar-height", "".concat(toolbarHeight, "px")),
      className: cnCollapseFullscreen({
        container: !containerIsBody
      }, [cnMixPopoverAnimate({
        animate: animate
      })])
    }, React.createElement(PortalWithThemeConsumer, {
      ignoreClicksInsideRefs: [portalRef]
    }, React.createElement("div", {
      className: cnCollapseFullscreen('ToolbarWrapper')
    }, React.createElement(Toolbar, {
      className: cnCollapseFullscreen('Toolbar'),
      ref: toolbarRef,
      leftSide: leftSide,
      form: "brick",
      border: "bottom",
      rightSide: [rightSide, React.createElement(CollapseButton, {
        active: active,
        icon: CollapseFullscreenIcon,
        onClick: onFullscreen
      })]
    })), React.createElement("div", {
      className: cnCollapseFullscreen('ChildrenWrapper')
    }, children)));
  });
};
