// @consta/table/__internal__/src/components/Table/helpers.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import { useCreateAtom } from '@consta/uikit/__internal__/src/utils/state/useCreateAtom';
import { getElementSize } from '@consta/uikit/useResizeObserved';
import { useAtom } from '@reatom/npm-react';
import { createRef } from 'react';
import { useResizeObservedAtom } from "../../hooks/useResizeObservedAtom";
import { get, set } from "../../utils/object/get";
export var columnDefaultMinWidth = 80;
export var separatorWidth = 8;
export var separatorLargeWidth = 24;

var reduceSum = function reduceSum(previousValue, currentValue) {
  return previousValue + currentValue;
};

var getLastChildrenCount = function getLastChildrenCount(columns) {
  var count = 0;

  var traverse = function traverse(cols) {
    cols.forEach(function (item) {
      var _item$columns;

      if ((_item$columns = item.columns) !== null && _item$columns !== void 0 && _item$columns.length) {
        traverse(item.columns);
      } else {
        count++;
      }
    });
  };

  traverse(columns);
  return count;
};

var mapColumns = function mapColumns(columns, fn) {
  var stopRef = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {
    current: false
  };
  var index = [0];

  while (index[0] <= columns.length - 1 && get(columns, index)) {
    var _item$columns2, _item$columns3;

    if (stopRef.current) {
      break;
    }

    var currentIndex = _toConsumableArray(index);

    var _item = get(columns, currentIndex);

    fn(_item, currentIndex);

    if ((_item$columns2 = _item.columns) !== null && _item$columns2 !== void 0 && _item$columns2.length) {
      index = [].concat(_toConsumableArray(currentIndex), ['columns', 0]);
    }

    if (!((_item$columns3 = _item.columns) !== null && _item$columns3 !== void 0 && _item$columns3.length)) {
      var nextIndex = [].concat(_toConsumableArray(currentIndex.slice(0, -1)), [Number(currentIndex[currentIndex.length - 1]) + 1]);

      if (get(columns, nextIndex)) {
        index = nextIndex;
      } else {
        if (currentIndex.length === 1) {
          index = [Number(currentIndex[0]) + 1];
          continue;
        }

        var d = 1;

        while (true) {
          var upIndex = currentIndex.slice(0, -d * 2);

          var _nextIndex = [].concat(_toConsumableArray(upIndex.slice(0, -1)), [Number(upIndex[upIndex.length - 1]) + 1]);

          if (!upIndex.length) {
            break;
          }

          if (get(columns, _nextIndex)) {
            index = _nextIndex;
            break;
          }

          if (_nextIndex[0] > columns.length - 1) {
            index = _nextIndex;
            break;
          }

          d += 1;
        }
      }
    }
  }
};

var pushByKey = function pushByKey(pushed, inColumns) {
  var needAdd = true;
  var stopRef = {
    current: false
  };
  mapColumns(inColumns, function (item) {
    if (pushed.key === item.key) {
      needAdd = false;
      stopRef.current = true;
    }
  }, stopRef);

  if (!needAdd) {
    return;
  }

  var keySplited = pushed.key.split('-');
  var parentKey = keySplited.slice(0, keySplited.length - 1).join('-');

  if (!parentKey) {
    inColumns.push(pushed);
    return;
  }

  mapColumns(inColumns, function (item, index) {
    if (parentKey === item.key) {
      var pushIndex = [].concat(_toConsumableArray(index), ['columns']);
      set(inColumns, pushIndex, [].concat(_toConsumableArray(get(inColumns, pushIndex)), [pushed]));
    }
  });
};

var pushByIndex = function pushByIndex(index, columns, inColumns, pinned) {
  var i = index.length - 1;

  while (i >= 0) {
    var currentI = i;
    i--;
    var tree = index.slice(0, index.length - currentI);

    if (tree[tree.length - 1] === 'columns') {
      continue;
    }

    pushByKey(_objectSpread(_objectSpread({}, get(columns, tree)), {}, {
      columns: [],
      pinned: pinned,
      key: tree.filter(function (el) {
        return el !== 'columns';
      }).join('-')
    }), inColumns);
  }
};

var transformPinnedColumns = function transformPinnedColumns(columns) {
  var pinnedLeftColumns = [];
  var pinnedRightColumns = [];
  var otherColumns = [];
  var notPinnedColumns = [];
  mapColumns(columns, function (item, index) {
    var _item$columns4, _item$columns5, _item$columns6;

    if (!((_item$columns4 = item.columns) !== null && _item$columns4 !== void 0 && _item$columns4.length) && item.pinned === 'left') {
      pushByIndex(index, columns, pinnedLeftColumns, 'left');
      return;
    }

    if (!((_item$columns5 = item.columns) !== null && _item$columns5 !== void 0 && _item$columns5.length) && item.pinned === 'right') {
      pushByIndex(index, columns, pinnedRightColumns, 'right');
      return;
    }

    if ((_item$columns6 = item.columns) !== null && _item$columns6 !== void 0 && _item$columns6.length || item.isSeparator || item.accessor) {
      pushByIndex(index, columns, otherColumns);
    }
  });
  mapColumns(otherColumns, function (item) {
    var _item$columns7;

    if ((_item$columns7 = item.columns) !== null && _item$columns7 !== void 0 && _item$columns7.length || item.isSeparator || item.accessor) {
      pushByKey(item, notPinnedColumns);
    }
  });
  return [].concat(pinnedLeftColumns, notPinnedColumns, pinnedRightColumns);
};

export var transformColumns = function transformColumns(columns, maxLevel) {
  var stack = [{
    columns: columns,
    index: 0
  }];
  var headersArr = [];
  var col = 0;

  while (stack.length) {
    var level = stack.length - 1;
    var node = stack[level];
    var _item2 = node.columns[node.index];

    if (_item2) {
      (function (_item2$colId, _handledItem$columns) {
        if (!headersArr[level]) headersArr[level] = [];
        var topHeaderGridIndex = stack[0].index;
        var prevItem = headersArr[level][headersArr[level].length - 1];
        var gridIndex = prevItem ? prevItem.position.gridIndex + (prevItem.position.colSpan || 1) : 0;
        var mainId = level === 0 ? col++ : (_item2$colId = _item2.colId) !== null && _item2$colId !== void 0 ? _item2$colId : 0;

        var handledItem = _objectSpread(_objectSpread({}, _item2), {}, {
          position: {
            topHeaderGridIndex: topHeaderGridIndex,
            gridIndex: gridIndex,
            level: level
          }
        });

        if (level === 0) {
          handledItem.colId = mainId;
        }

        if (!((_handledItem$columns = handledItem.columns) !== null && _handledItem$columns !== void 0 && _handledItem$columns.length)) {
          handledItem.position.rowSpan = maxLevel - level;
          headersArr[level].push(handledItem);
          node.index++;
        } else {
          handledItem.position.colSpan = getLastChildrenCount(handledItem.columns);
          headersArr[level].push(handledItem);
          stack.push({
            columns: handledItem.columns.map(function (el) {
              return _objectSpread(_objectSpread({}, el), {}, {
                colId: col++,
                parentId: mainId
              });
            }),
            index: 0
          });
        }
      })();
    } else {
      stack.pop();
      if (stack[stack.length - 1]) stack[stack.length - 1].index++;
    }
  }

  return headersArr;
};
export var getMaxLevel = function getMaxLevel(columns) {
  var count = 0;

  var traverse = function traverse(cols) {
    var level = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
    if (level > count) count = level;
    cols.forEach(function (item) {
      var _item$columns8;

      if ((_item$columns8 = item.columns) !== null && _item$columns8 !== void 0 && _item$columns8.length) {
        traverse(item.columns, level + 1);
      }
    });
  };

  traverse(columns);
  return count;
};

var getIsFirst = function getIsFirst(columns, column) {
  var _parent$columns, _parent$columns$;

  var colId = column.colId,
      parentId = column.parentId,
      position = column.position,
      accessor = column.accessor;

  if (position.level === 0) {
    return colId === 0;
  }

  var parent = columns.find(function (el) {
    return el.colId === parentId;
  });
  return !!((parent === null || parent === void 0 ? void 0 : (_parent$columns = parent.columns) === null || _parent$columns === void 0 ? void 0 : (_parent$columns$ = _parent$columns[0]) === null || _parent$columns$ === void 0 ? void 0 : _parent$columns$.accessor) === accessor && (parent ? getIsFirst(columns, parent) : false));
};

var getLowHeaders = function getLowHeaders(columns) {
  var lowHeaders = [];
  mapColumns(columns, function (item) {
    var _item$columns9;

    if (!((_item$columns9 = item.columns) !== null && _item$columns9 !== void 0 && _item$columns9.length)) {
      lowHeaders.push(item);
    }
  });
  return lowHeaders;
};

var getStickyTopOffsets = function getStickyTopOffsets(flattenedHeaders, headerRowsHeights) {
  var stickyTopOffsets = [];

  for (var index = 0; index < flattenedHeaders.length; index++) {
    var column = flattenedHeaders[index];
    stickyTopOffsets.push(headerRowsHeights.slice(0, column.position.level).reduce(reduceSum, 0));
  }

  return stickyTopOffsets;
};

var getResizerTopOffsets = function getResizerTopOffsets(flattenedHeaders, lowHeaders, stickyTopOffsets) {
  var resizerTopOffsets = [];

  var _loop = function _loop(index) {
    var topOffsetIndex = flattenedHeaders.findIndex(function (item) {
      return item.key === lowHeaders[index].key;
    });
    resizerTopOffsets.push(stickyTopOffsets[topOffsetIndex]);
  };

  for (var index = 0; index < lowHeaders.length; index++) {
    _loop(index);
  }

  return resizerTopOffsets;
};

var getFlattenedHeadersLowCellsKeys = function getFlattenedHeadersLowCellsKeys(flattenedHeaders) {
  var lowCells = flattenedHeaders.map(function (flattenedHeaderCell) {
    var keys = [];
    mapColumns([flattenedHeaderCell], function (item, index) {
      var _item$columns10;

      if (!((_item$columns10 = item.columns) !== null && _item$columns10 !== void 0 && _item$columns10.length)) {
        keys.push(item.key);
      }
    });
    return keys;
  });
  return lowCells;
};

var getHeaderStickyLeftOffsets = function getHeaderStickyLeftOffsets(lowCellsKeys, lowHeaders) {
  return lowCellsKeys.map(function (keys) {
    return Math.min.apply(Math, _toConsumableArray(keys.map(function (key) {
      return lowHeaders.findIndex(function (item) {
        return item.key === key;
      });
    })));
  });
};

var getHeaderStickyRightOffsets = function getHeaderStickyRightOffsets(lowCellsKeys, lowHeaders) {
  return lowCellsKeys.map(function (keys) {
    return Math.max.apply(Math, _toConsumableArray(keys.map(function (key) {
      return lowHeaders.findIndex(function (item) {
        return item.key === key;
      });
    })));
  });
};

export var useHeaderData = function useHeaderData(columnsAtom, virtualScrollAtom) {
  var horizontalVirtualScrollAtom = useCreateAtom(function (ctx) {
    var virtualScroll = ctx.spy(virtualScrollAtom) || false;
    return Array.isArray(virtualScroll) ? virtualScroll[0] : virtualScroll;
  });
  var columnsWithPinnedAtom = useCreateAtom(function (ctx) {
    return transformPinnedColumns(ctx.spy(columnsAtom));
  });
  var headersAtom = useCreateAtom(function (ctx) {
    var columnsWithPinned = ctx.spy(columnsWithPinnedAtom);
    return transformColumns(columnsWithPinned, getMaxLevel(columnsWithPinned));
  });
  var flattenedHeadersAtom = useCreateAtom(function (ctx) {
    var headers = ctx.spy(headersAtom);
    return headers.flat().map(function (column, index, array) {
      return _objectSpread(_objectSpread({}, column), {}, {
        position: _objectSpread(_objectSpread({}, column.position), {}, {
          isFirst: getIsFirst(array, column),
          width: column.width || 'auto'
        })
      });
    });
  });
  var flattenedHeadersLengthAtom = useCreateAtom(function (ctx) {
    return ctx.spy(flattenedHeadersAtom).length;
  });
  var headerCellsRefsAtom = useCreateAtom(function (ctx) {
    return new Array(ctx.spy(flattenedHeadersLengthAtom)).fill(null).map(createRef);
  });
  var headerCellsHeightsAtom = useResizeObservedAtom(useAtom(headerCellsRefsAtom)[0], function (el) {
    return getElementSize(el).height;
  });
  var headerCellsHeightsHashAtom = useCreateAtom(function (ctx) {
    var headerCellsHeights = ctx.spy(headerCellsHeightsAtom);
    return headerCellsHeights.join('-');
  });
  var headerRowsHeightsAtom = useCreateAtom(function (ctx) {
    var headers = ctx.spy(headersAtom);
    var flattenedHeaders = ctx.spy(flattenedHeadersAtom);
    ctx.spy(headerCellsHeightsHashAtom);
    var headerCellsHeights = ctx.get(headerCellsHeightsAtom);
    return headers.map(function (arr, index) {
      var flattenedHeadersWithHeights = flattenedHeaders.map(function (item, i) {
        return _objectSpread(_objectSpread({}, item), {}, {
          position: _objectSpread(_objectSpread({}, item.position), {}, {
            height: headerCellsHeights[i]
          })
        });
      });
      return Math.min.apply(null, flattenedHeadersWithHeights.filter(function (col) {
        return col.position.level === index;
      }).map(function (item, i) {
        return item.position.height;
      }));
    });
  });
  var lowHeadersAtom = useCreateAtom(function (ctx) {
    return getLowHeaders(ctx.spy(columnsWithPinnedAtom));
  });
  var lowHeaderslengthAtom = useCreateAtom(function (ctx) {
    return ctx.spy(lowHeadersAtom).length;
  });
  var resizersRefsAtom = useCreateAtom(function (ctx) {
    return new Array(ctx.spy(lowHeaderslengthAtom)).fill(null).map(createRef);
  });
  var headerHeightAtom = useCreateAtom(function (ctx) {
    return ctx.spy(headerRowsHeightsAtom).reduce(reduceSum);
  });
  var stickyTopOffsetsAtom = useCreateAtom(function (ctx) {
    return getStickyTopOffsets(ctx.spy(flattenedHeadersAtom), ctx.spy(headerRowsHeightsAtom));
  });
  var resizerTopOffsetsAtom = useCreateAtom(function (ctx) {
    return getResizerTopOffsets(ctx.spy(flattenedHeadersAtom), ctx.spy(lowHeadersAtom), ctx.spy(stickyTopOffsetsAtom));
  });
  var flattenedHeadersLowCellsKeysAtom = useCreateAtom(function (ctx) {
    return getFlattenedHeadersLowCellsKeys(ctx.spy(flattenedHeadersAtom));
  });
  var stickyLeftOffsetsAtom = useCreateAtom(function (ctx) {
    return getHeaderStickyLeftOffsets(ctx.spy(flattenedHeadersLowCellsKeysAtom), ctx.spy(lowHeadersAtom));
  });
  var stickyRightOffsetsAtom = useCreateAtom(function (ctx) {
    return getHeaderStickyRightOffsets(ctx.spy(flattenedHeadersLowCellsKeysAtom), ctx.spy(lowHeadersAtom));
  });
  var bordersFlattenedHeadersAtom = useCreateAtom(function (ctx) {
    var flattenedHeaders = ctx.spy(flattenedHeadersAtom);
    var lowHeaders = ctx.spy(lowHeadersAtom);
    return flattenedHeaders.map(function (flattenedHeadersColumn, index) {
      var _lowHeaders, _flattenedHeaders;

      var prevLowKey = '';
      var stopRef = {
        current: false
      };
      mapColumns([flattenedHeadersColumn], function (col) {
        var _col$columns;

        if (!((_col$columns = col.columns) !== null && _col$columns !== void 0 && _col$columns.length)) {
          prevLowKey = col.key;
          stopRef.current = true;
        }
      }, stopRef);
      var lowIndex = lowHeaders.findIndex(function (col) {
        return col.key === prevLowKey;
      });
      return [!flattenedHeadersColumn.position.isFirst && !(flattenedHeadersColumn.pinned !== 'left' && ((_lowHeaders = lowHeaders[lowIndex - 1]) === null || _lowHeaders === void 0 ? void 0 : _lowHeaders.pinned) === 'left'), flattenedHeadersColumn.pinned === 'left' && ((_flattenedHeaders = flattenedHeaders[index + 1]) === null || _flattenedHeaders === void 0 ? void 0 : _flattenedHeaders.pinned) !== 'left', flattenedHeadersColumn.position.level !== 0];
    });
  });
  var intersectingColumnsAtom = useCreateAtom([]);
  var leftNoVisibleItemsAtom = useCreateAtom(function (ctx) {
    var intersectingColumns = ctx.spy(intersectingColumnsAtom);
    var horizontalVirtualScroll = ctx.spy(horizontalVirtualScrollAtom);

    if (!horizontalVirtualScroll) {
      return 0;
    }

    if (intersectingColumns.length === 0) {
      return ctx.get(lowHeadersAtom).length;
    }

    var offset = 0;

    while (intersectingColumns[offset] === false) {
      offset++;
    }

    return offset;
  });
  var rightNoVisibleItemsAtom = useCreateAtom(function (ctx) {
    var intersectingColumns = ctx.spy(intersectingColumnsAtom);
    var horizontalVirtualScroll = ctx.spy(horizontalVirtualScrollAtom);

    if (!horizontalVirtualScroll) {
      return 0;
    }

    var offset = intersectingColumns.length - 1;

    while (intersectingColumns[offset] === false) {
      offset--;
    }

    return intersectingColumns.length - offset - 1;
  });
  return {
    headersAtom: headersAtom,
    flattenedHeadersAtom: flattenedHeadersAtom,
    lowHeadersAtom: lowHeadersAtom,
    headerRowsHeightsAtom: headerRowsHeightsAtom,
    resizerTopOffsetsAtom: resizerTopOffsetsAtom,
    headerHeightAtom: headerHeightAtom,
    resizersRefsAtom: resizersRefsAtom,
    stickyTopOffsetsAtom: stickyTopOffsetsAtom,
    stickyLeftOffsetsAtom: stickyLeftOffsetsAtom,
    stickyRightOffsetsAtom: stickyRightOffsetsAtom,
    headerCellsRefsAtom: headerCellsRefsAtom,
    bordersFlattenedHeadersAtom: bordersFlattenedHeadersAtom,
    intersectingColumnsAtom: intersectingColumnsAtom,
    rightNoVisibleItemsAtom: rightNoVisibleItemsAtom,
    leftNoVisibleItemsAtom: leftNoVisibleItemsAtom
  };
};
