// @consta/table/__internal__/src/components/Table/Table.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["columns", "rows", "stickyHeader", "virtualScroll", "resizable", "zebraStriped", "headerZIndex", "onRowMouseEnter", "onRowMouseLeave", "onRowClick", "getRowKey", "rowHoverEffect", "onScrollToBottom"];
import { objectWithDefault } from '@consta/uikit/__internal__/src/utils/object/objectWithDefault';
import { usePropAtom } from '@consta/uikit/__internal__/src/utils/state/usePickAtom';
import { useSendToAtom } from '@consta/uikit/__internal__/src/utils/state/useSendToAtom';
import { withCtx } from '@consta/uikit/__internal__/src/utils/state/withCtx';
import { useForkRef } from '@consta/uikit/useForkRef';
import React, { forwardRef, memo } from 'react';
import { useVirtualScrollAtom } from "../../hooks/useVirtualScrollAtom";
import { useHeaderData } from "./helpers";
import { TableBody } from "./TableBody";
import { TableData } from "./TableData";
import { TableHeader } from "./TableHeader/TableHeader";
var MemoTableHeader = memo(TableHeader);
var MemoTableData = memo(TableData);
export var TableRender = function TableRender(props, ref) {
  var columns = props.columns,
      rows = props.rows,
      stickyHeader = props.stickyHeader,
      virtualScroll = props.virtualScroll,
      resizable = props.resizable,
      zebraStriped = props.zebraStriped,
      _props$headerZIndex = props.headerZIndex,
      headerZIndex = _props$headerZIndex === void 0 ? 1 : _props$headerZIndex,
      onRowMouseEnter = props.onRowMouseEnter,
      onRowMouseLeave = props.onRowMouseLeave,
      onRowClick = props.onRowClick,
      getRowKey = props.getRowKey,
      rowHoverEffect = props.rowHoverEffect,
      onScrollToBottom = props.onScrollToBottom,
      otherProps = _objectWithoutProperties(props, _excluded);

  var propsAtom = useSendToAtom(objectWithDefault({
    headerZIndex: 1
  }, props));

  var _useHeaderData = useHeaderData(usePropAtom(propsAtom, 'columns'), usePropAtom(propsAtom, 'virtualScroll')),
      resizerTopOffsetsAtom = _useHeaderData.resizerTopOffsetsAtom,
      headerHeightAtom = _useHeaderData.headerHeightAtom,
      resizersRefsAtom = _useHeaderData.resizersRefsAtom,
      lowHeadersAtom = _useHeaderData.lowHeadersAtom,
      stickyTopOffsetsAtom = _useHeaderData.stickyTopOffsetsAtom,
      flattenedHeadersAtom = _useHeaderData.flattenedHeadersAtom,
      headerCellsRefsAtom = _useHeaderData.headerCellsRefsAtom,
      stickyLeftOffsetsAtom = _useHeaderData.stickyLeftOffsetsAtom,
      stickyRightOffsetsAtom = _useHeaderData.stickyRightOffsetsAtom,
      bordersFlattenedHeadersAtom = _useHeaderData.bordersFlattenedHeadersAtom,
      intersectingColumnsAtom = _useHeaderData.intersectingColumnsAtom,
      leftNoVisibleItemsAtom = _useHeaderData.leftNoVisibleItemsAtom,
      rightNoVisibleItemsAtom = _useHeaderData.rightNoVisibleItemsAtom;

  var _useVirtualScrollAtom = useVirtualScrollAtom({
    length: rows.length,
    isActive: Array.isArray(virtualScroll) ? virtualScroll[1] : virtualScroll,
    onScrollToBottom: onScrollToBottom
  }),
      listRefsAtom = _useVirtualScrollAtom.listRefsAtom,
      scrollElementRef = _useVirtualScrollAtom.scrollElementRef,
      sliceAtom = _useVirtualScrollAtom.sliceAtom,
      spaceTopAtom = _useVirtualScrollAtom.spaceTopAtom;

  return React.createElement(TableBody, Object.assign({}, otherProps, {
    topOffsetsAtom: resizerTopOffsetsAtom,
    spaceTopAtom: spaceTopAtom,
    ref: useForkRef([scrollElementRef, ref]),
    headerHeightAtom: headerHeightAtom,
    lowHeadersAtom: lowHeadersAtom,
    resizersRefsAtom: resizersRefsAtom,
    resizable: resizable,
    stickyTopOffsetsAtom: stickyTopOffsetsAtom,
    stickyHeader: stickyHeader,
    headerZIndex: headerZIndex,
    intersectingColumnsAtom: intersectingColumnsAtom,
    header: React.createElement(MemoTableHeader, {
      headersAtom: flattenedHeadersAtom,
      headerCellsRefsAtom: headerCellsRefsAtom,
      stickyHeader: stickyHeader,
      stickyLeftOffsetsAtom: stickyLeftOffsetsAtom,
      stickyRightOffsetsAtom: stickyRightOffsetsAtom,
      bordersAtom: bordersFlattenedHeadersAtom,
      tableRef: scrollElementRef
    }),
    body: React.createElement(MemoTableData, {
      lowHeadersAtom: lowHeadersAtom,
      rows: rows,
      rowsRefsAtom: listRefsAtom,
      sliceAtom: sliceAtom,
      zebraStriped: zebraStriped,
      onRowMouseEnter: onRowMouseEnter,
      onRowMouseLeave: onRowMouseLeave,
      onRowClick: onRowClick,
      getRowKey: getRowKey,
      tableRef: scrollElementRef,
      rowHoverEffect: rowHoverEffect,
      leftNoVisibleItemsAtom: leftNoVisibleItemsAtom,
      rightNoVisibleItemsAtom: rightNoVisibleItemsAtom
    })
  }));
};
export var Table = withCtx(forwardRef(TableRender));
