// @consta/table/__internal__/src/components/Table/TableRowCell/TableRowCell.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["className", "index", "separator", "accessor", "pinned", "borderLeft", "borderRight", "borderTop", "miss", "renderCell", "row", "rowIndex", "tableRef"];
import "./TableRowCell.css";
import React, { forwardRef, memo } from 'react';
import { DataCell } from "../../DataCell";
import { cn } from "../../../utils/bem";
import { isNotNil, isNumber, isString } from "../../../utils/type-guards";
import { cnTableCell } from "../TableCell";
var cnTableRowCell = cn('TableRowCell');

var getCellDataByAccessor = function getCellDataByAccessor(row, accessor, separator) {
  if (separator) {
    return '';
  }

  if (!accessor) {
    return '';
  }

  var data = row === null || row === void 0 ? void 0 : row[accessor];

  if (isString(data) || isNumber(data)) {
    return React.createElement(DataCell, null, data.toString());
  }

  return '';
};

var TableRowCellRender = function TableRowCellRender(props, ref) {
  var className = props.className,
      index = props.index,
      separator = props.separator,
      accessor = props.accessor,
      pinned = props.pinned,
      borderLeft = props.borderLeft,
      borderRight = props.borderRight,
      borderTop = props.borderTop,
      miss = props.miss,
      RenderCell = props.renderCell,
      row = props.row,
      rowIndex = props.rowIndex,
      tableRef = props.tableRef,
      otherProps = _objectWithoutProperties(props, _excluded);

  return React.createElement("div", Object.assign({}, otherProps, {
    ref: ref,
    className: cnTableRowCell({
      pinned: !!pinned
    }, [cnTableCell({
      separator: separator,
      borderLeft: borderLeft,
      borderRight: borderRight,
      borderTop: borderTop,
      sticky: !!pinned
    }), className]),
    style: {
      left: pinned === 'left' ? "var(--table-column-sticky-left-offset-".concat(index, ")") : undefined,
      right: pinned === 'right' ? "var(--table-column-sticky-right-offset-".concat(index, ")") : undefined,
      gridColumn: "".concat(index + 1, " / span ").concat(miss > 0 ? miss + 1 : 1)
    }
  }), isNotNil(RenderCell) ? React.createElement(RenderCell, {
    row: row,
    rowIndex: rowIndex,
    columnIndex: index,
    tableRef: tableRef
  }) : getCellDataByAccessor(row, accessor, separator));
};

export var TableRowCell = memo(forwardRef(TableRowCellRender));
