// @consta/table/__internal__/src/utils/array.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

export var sortBy = function sortBy(array, param) {
  var order = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'asc';
  var sortFn = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : function (a, b) {
    return a < b ? -1 : 1;
  };

  var compareFn = function compareFn(a, b) {
    return (order === 'desc' ? -1 : 1) * sortFn(a[param], b[param]);
  };

  return _toConsumableArray(array).sort(compareFn);
};
export var updateAt = function updateAt(array, index, newItem) {
  var isExpand = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
  var fixedIndex = index < 0 ? 0 : index;

  if (fixedIndex >= array.length) {
    return isExpand ? [].concat(_toConsumableArray(array), [newItem]) : array;
  }

  return [].concat(_toConsumableArray(array.slice(0, fixedIndex)), [newItem], _toConsumableArray(array.slice(fixedIndex + (isExpand ? 0 : 1), array.length)));
};
export var putBefore = function putBefore(array, index, newItem) {
  return updateAt(array, index, newItem, true);
};
export var range = function range(n) {
  return _toConsumableArray(Array(n).keys());
};
export var chunk = function chunk(array, size) {
  var index = 0;
  var resIndex = 0;
  var result = [];

  while (index < array.length) {
    result[resIndex++] = array.slice(index, index += size);
  }

  return result;
};
export var groupBy = function groupBy(array, iteratee) {
  return array.reduce(function (result, value) {
    var key = iteratee(value);
    return _objectSpread(_objectSpread({}, result), {}, _defineProperty({}, key, result[key] ? [].concat(_toConsumableArray(result[key]), [value]) : [value]));
  }, {});
};
