// @consta/table/__internal__/src/utils/cnForCssTransition.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function cnFunctionHelper(cnFunction, animateName, animate, element) {
  return element ? cnFunction(element, _defineProperty({}, animateName, animate)) : cnFunction(_defineProperty({}, animateName, animate));
}

export function cnForCssTransition(cnFunction, element) {
  var animateName = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'animate';
  return {
    appear: cnFunctionHelper(cnFunction, animateName, 'appear', element),
    appearActive: cnFunctionHelper(cnFunction, animateName, 'appearActive', element),
    appearDone: cnFunctionHelper(cnFunction, animateName, 'appearDone', element),
    enter: cnFunctionHelper(cnFunction, animateName, 'enter', element),
    enterActive: cnFunctionHelper(cnFunction, animateName, 'enterActive', element),
    enterDone: cnFunctionHelper(cnFunction, animateName, 'enterDone', element),
    exit: cnFunctionHelper(cnFunction, animateName, 'exit', element),
    exitActive: cnFunctionHelper(cnFunction, animateName, 'exitActive', element),
    exitDone: cnFunctionHelper(cnFunction, animateName, 'exitDone', element)
  };
}
