// @consta/table/__internal__/src/utils/getGroups.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

export var defaultGroupKey = 'no-group';

var getItemGroupKeyGuard = function getItemGroupKeyGuard(getter, item, noGroupKey) {
  var key = getter(item);
  return typeof key === 'undefined' ? noGroupKey : key;
};

export function getGroups(items, getItemGroupKey, groups, getGroupKey, sortGroups) {
  var noGroupKey = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : defaultGroupKey;

  if (typeof getItemGroupKey !== 'function') {
    return [{
      items: items,
      key: noGroupKey,
      groupIndex: -1
    }];
  }

  var resultGroups = [];

  var _iterator = _createForOfIteratorHelper(items),
      _step;

  try {
    var _loop = function _loop() {
      var item = _step.value;
      var itemGroupKey = getItemGroupKeyGuard(getItemGroupKey, item, noGroupKey);
      var resultGroupIndex = resultGroups.findIndex(function (group) {
        return group.key === itemGroupKey;
      });

      if (resultGroupIndex >= 0) {
        resultGroups[resultGroupIndex].items.push(item);
      } else {
        var groupIndex = getGroupKey && groups ? groups.findIndex(function (g) {
          return getGroupKey(g) === itemGroupKey;
        }) : -1;
        var resultGroup = {
          key: itemGroupKey,
          items: [item],
          groupIndex: groupIndex,
          group: groups ? groups[groupIndex] : undefined
        };
        resultGroups.push(resultGroup);
      }
    };

    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      _loop();
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  if (typeof sortGroups === 'function') {
    resultGroups.sort(sortGroups);
  } else if (groups && getGroupKey) {
    resultGroups.sort(function (a, b) {
      if (a.groupIndex > b.groupIndex) return 1;
      if (a.groupIndex < b.groupIndex) return -1;
      return 0;
    });
  }

  return resultGroups;
}
export function getCountedGroups(groups, values, selectAll, getItemKey, getItemDisabled) {
  var copyGroups = _toConsumableArray(groups);

  if (selectAll) {
    groups.forEach(function (group, index) {
      var totalCount = 0;
      var checkedCount = 0;
      group.items.forEach(function (item) {
        if (!(getItemDisabled && getItemDisabled !== null && getItemDisabled !== void 0 && getItemDisabled(item))) {
          totalCount += 1;
        }

        if (values !== null && values !== void 0 && values.find(function (el) {
          return getItemKey(item) === getItemKey(el);
        })) {
          checkedCount += 1;
        }
      });
      copyGroups[index].items = [{
        __optionSelctAll: true,
        totalCount: totalCount,
        groupKey: group.key,
        checkedCount: checkedCount
      }].concat(_toConsumableArray(copyGroups[index].items));
    });
  }

  return copyGroups;
}
