// @consta/table/__internal__/src/components/Collapse/Collapse.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["expandButton", "expanded", "onExpand", "fullscreenButton", "fullscreen", "onFullscreen", "leftSide", "rightSide", "className", "children", "fullscreenContainer", "fullscreenZIndex", "expandedMaxHeight"];
import "./Collapse.css";
import { withCtx } from '@consta/uikit/__internal__/src/utils/state/withCtx';
import { isString } from '@consta/uikit/__internal__/src/utils/type-guards';
import { animateTimeout } from '@consta/uikit/MixPopoverAnimate';
import { Text } from '@consta/uikit/Text';
import { useMutableRef } from '@consta/uikit/useMutableRef';
import { useAction, useAtom, useUpdate } from '@reatom/npm-react';
import React, { forwardRef, useRef } from 'react';
import { Transition } from 'react-transition-group';
import { Toolbar } from "../Toolbar";
import { cn } from "../../utils/bem";
import { CollapseButton } from "./CollapseButton";
import { CollapseExpandIcon } from "./CollapseExpandIcon";
import { CollapseFullscreen } from "./CollapseFullscreen";
import { CollapseFullscreenIcon } from "./CollapseFullscreenIcon";
var cnCollapse = cn('Collapse');
export var Collapse = withCtx(forwardRef(function (props, ref) {
  var expandButton = props.expandButton,
      _props$expanded = props.expanded,
      expandedProp = _props$expanded === void 0 ? false : _props$expanded,
      onExpand = props.onExpand,
      fullscreenButton = props.fullscreenButton,
      _props$fullscreen = props.fullscreen,
      fullscreenProp = _props$fullscreen === void 0 ? false : _props$fullscreen,
      onFullscreen = props.onFullscreen,
      leftSide = props.leftSide,
      rightSide = props.rightSide,
      className = props.className,
      children = props.children,
      _props$fullscreenCont = props.fullscreenContainer,
      fullscreenContainer = _props$fullscreenCont === void 0 ? window.document.body : _props$fullscreenCont,
      _props$fullscreenZInd = props.fullscreenZIndex,
      fullscreenZIndex = _props$fullscreenZInd === void 0 ? 1000 : _props$fullscreenZInd,
      _props$expandedMaxHei = props.expandedMaxHeight,
      expandedMaxHeight = _props$expandedMaxHei === void 0 ? 'auto' : _props$expandedMaxHei,
      otherProps = _objectWithoutProperties(props, _excluded);

  var refs = useMutableRef([onExpand, onFullscreen]);
  var contentRef = useRef(null);

  var _useAtom = useAtom(expandedProp),
      _useAtom2 = _slicedToArray(_useAtom, 3),
      expanded = _useAtom2[0],
      expandedAtom = _useAtom2[2];

  var _useAtom3 = useAtom(fullscreenProp),
      _useAtom4 = _slicedToArray(_useAtom3, 3),
      fullscreen = _useAtom4[0],
      fullscreenAtom = _useAtom4[2];

  var _useAtom5 = useAtom(undefined),
      _useAtom6 = _slicedToArray(_useAtom5, 2),
      fakeContentHeight = _useAtom6[0],
      setFakeContentHeight = _useAtom6[1];

  var _useAtom7 = useAtom(expanded),
      _useAtom8 = _slicedToArray(_useAtom7, 3),
      expandedContentMounted = _useAtom8[0],
      expandedContentMountedAtom = _useAtom8[2];

  var toggleExpanded = useAction(function (ctx, e) {
    var _refs$current$, _refs$current;

    var value = !ctx.get(expandedAtom);
    (_refs$current$ = (_refs$current = refs.current)[0]) === null || _refs$current$ === void 0 ? void 0 : _refs$current$.call(_refs$current, value, e);
    expandedAtom(ctx, value);

    if (value) {
      expandedContentMountedAtom(ctx, true);
    }
  });
  var toggleFullscreen = useAction(function (ctx, e) {
    var _refs$current$2, _refs$current2, _contentRef$current;

    var value = !ctx.get(fullscreenAtom);
    (_refs$current$2 = (_refs$current2 = refs.current)[1]) === null || _refs$current$2 === void 0 ? void 0 : _refs$current$2.call(_refs$current2, value, e);
    fullscreenAtom(ctx, value);
    setFakeContentHeight(value ? (_contentRef$current = contentRef.current) === null || _contentRef$current === void 0 ? void 0 : _contentRef$current.clientHeight : undefined);
  });
  var afterExitExpanded = useAction(function (ctx) {
    expandedContentMountedAtom(ctx, false);
  });
  var title = isString(leftSide) ? React.createElement(Text, {
    className: cnCollapse('Title'),
    size: "m",
    weight: "semibold"
  }, leftSide) : leftSide;
  useUpdate(expandedAtom, [expandedProp]);
  useUpdate(fullscreenAtom, [fullscreenProp]);
  return React.createElement(React.Fragment, null, React.createElement("div", Object.assign({}, otherProps, {
    ref: ref,
    className: className
  }), React.createElement(Toolbar, {
    className: cnCollapse('Toolbar'),
    form: "brick",
    border: expandedContentMounted ? 'bottom' : undefined,
    leftSide: [expandButton ? React.createElement(CollapseButton, {
      active: expanded,
      icon: CollapseExpandIcon,
      onClick: toggleExpanded
    }) : null].concat(_toConsumableArray(Array.isArray(title) ? title : [title])),
    rightSide: [].concat(_toConsumableArray(Array.isArray(rightSide) ? rightSide : [rightSide]), [fullscreenButton ? React.createElement(CollapseButton, {
      active: fullscreen,
      icon: CollapseFullscreenIcon,
      onClick: toggleFullscreen
    }) : null])
  }), children && React.createElement("div", {
    className: cnCollapse('Content', {
      expanded: expanded
    }),
    style: {
      height: fakeContentHeight
    }
  }, React.createElement(Transition, {
    "in": expanded && !fullscreen,
    unmountOnExit: true,
    nodeRef: contentRef,
    timeout: animateTimeout,
    onExited: afterExitExpanded
  }, React.createElement("div", {
    className: cnCollapse('ChildrenWrapper'),
    ref: contentRef,
    style: {
      maxHeight: expandedMaxHeight
    }
  }, children)))), children && React.createElement(CollapseFullscreen, {
    active: fullscreen,
    onFullscreen: toggleFullscreen,
    container: fullscreenContainer,
    zIndex: fullscreenZIndex,
    leftSide: title,
    rightSide: rightSide
  }, children));
}));
