// @consta/table/__internal__/src/components/Table/TableBody/useResizableColumns/useResizableColumns.js
import { useCreateAtom } from '@consta/uikit/__internal__/src/utils/state/useCreateAtom';
import { useSendToAtom } from '@consta/uikit/__internal__/src/utils/state/useSendToAtom';
import { useAction, useUpdate } from '@reatom/npm-react';
import { useEffect } from 'react';
import { addResult, getCalculatedSizes, getRefsSizes, sizesEq, useResizeContainer } from "./helpers";
export var useResizableColumns = function useResizableColumns(props) {
  var propsAtom = useSendToAtom(props);
  var blocksLengthAtom = useCreateAtom(function (ctx) {
    return ctx.get(propsAtom).blocks.length;
  });
  var blocksAtom = useCreateAtom(function (ctx) {
    return ctx.spy(propsAtom).blocks;
  });
  var resizableAtom = useCreateAtom(function (ctx) {
    return ctx.spy(propsAtom).resizable;
  });
  var containerAtom = useCreateAtom(function (ctx) {
    return ctx.spy(propsAtom).container;
  });
  var blocks = props.blocks;
  var sizesAtom = useCreateAtom(getRefsSizes(blocks));
  var activeIndexAtom = useCreateAtom(null);
  var resizingAtom = useCreateAtom(function (ctx) {
    return typeof ctx.spy(activeIndexAtom) === 'number';
  });
  var setSizes = useAction(function (ctx, newSizes) {
    if (!sizesEq(newSizes, ctx.get(sizesAtom))) {
      sizesAtom(ctx, newSizes);
    }
  });
  var controlListeners = useAction(function (_, type) {
    var method = type === 'add' ? 'addEventListener' : 'removeEventListener';
    document[method]('mouseup', handleRelease);
    document[method]('touchend', handleRelease);
    document[method]('mousemove', handleTouchMove);
    document[method]('touchmove', handleTouchMove);
  });
  var handleRelease = useAction(function (ctx) {
    activeIndexAtom(ctx, null);
    controlListeners('remove');
  });
  var handlePress = useAction(function (ctx, index) {
    activeIndexAtom(ctx, index);
    controlListeners('add');
  });
  var handleTouchMove = useAction(function (ctx, event) {
    var _ctx$get = ctx.get(propsAtom),
        container = _ctx$get.container,
        blocks = _ctx$get.blocks,
        resizable = _ctx$get.resizable;

    var activeIndex = ctx.get(activeIndexAtom);

    if (typeof activeIndex === 'number' && resizable) {
      var sizes = ctx.get(sizesAtom);
      setSizes(addResult(getCalculatedSizes(event, activeIndex, blocks, container, sizes, resizable), sizes));
    }
  });
  var handlersAtom = useCreateAtom(function (ctx) {
    var length = ctx.spy(blocksLengthAtom);
    var resizable = ctx.spy(resizableAtom);

    if (!resizable) {
      return [];
    }

    return Array.from({
      length: length
    }).map(function (_el, index) {
      return {
        onMouseDown: function onMouseDown() {
          return handlePress(index);
        },
        onTouchStart: function onTouchStart() {
          return handlePress(index);
        }
      };
    });
  });
  useResizeContainer(containerAtom, blocksAtom, resizableAtom, sizesAtom, setSizes);
  useUpdate(function (_, blocks) {
    setSizes(getRefsSizes(blocks));
  }, [blocksAtom, resizableAtom]);
  useUpdate(function (_, resizable) {
    !resizable && handleRelease();
  }, [resizableAtom]);
  useEffect(function () {
    return handleRelease;
  }, []);
  return {
    handlersAtom: handlersAtom,
    sizesAtom: sizesAtom,
    activeIndexAtom: activeIndexAtom,
    resizingAtom: resizingAtom
  };
};
